// react-native-render-html/packages/render-html/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  HTMLContentModel: true,
  TStyles: true,
  defaultHTMLElementModels: true,
  tnodeToString: true,
  TChildrenRenderer: true,
  TNodeChildrenRenderer: true,
  useTNodeChildrenProps: true,
  TNodeRenderer: true,
  useTRenderEngine: true,
  useTTree: true,
  defaultRenderers: true,
  extendDefaultRenderer: true,
  splitBoxModelStyle: true,
  useSharedProps: true
};
Object.defineProperty(exports, "HTMLContentModel", {
  enumerable: true,
  get: function () {
    return _transientRenderEngine.HTMLContentModel;
  }
});
Object.defineProperty(exports, "TStyles", {
  enumerable: true,
  get: function () {
    return _transientRenderEngine.TStyles;
  }
});
Object.defineProperty(exports, "defaultHTMLElementModels", {
  enumerable: true,
  get: function () {
    return _transientRenderEngine.defaultHTMLElementModels;
  }
});
Object.defineProperty(exports, "tnodeToString", {
  enumerable: true,
  get: function () {
    return _transientRenderEngine.tnodeToString;
  }
});
Object.defineProperty(exports, "TChildrenRenderer", {
  enumerable: true,
  get: function () {
    return _TChildrenRenderer.default;
  }
});
Object.defineProperty(exports, "TNodeChildrenRenderer", {
  enumerable: true,
  get: function () {
    return _TNodeChildrenRenderer.default;
  }
});
Object.defineProperty(exports, "useTNodeChildrenProps", {
  enumerable: true,
  get: function () {
    return _TNodeChildrenRenderer.useTNodeChildrenProps;
  }
});
Object.defineProperty(exports, "TNodeRenderer", {
  enumerable: true,
  get: function () {
    return _TNodeRenderer.default;
  }
});
Object.defineProperty(exports, "useTRenderEngine", {
  enumerable: true,
  get: function () {
    return _useTRenderEngine.default;
  }
});
Object.defineProperty(exports, "useTTree", {
  enumerable: true,
  get: function () {
    return _useTTree.default;
  }
});
Object.defineProperty(exports, "defaultRenderers", {
  enumerable: true,
  get: function () {
    return _defaultRenderers.default;
  }
});
Object.defineProperty(exports, "extendDefaultRenderer", {
  enumerable: true,
  get: function () {
    return _extendDefaultRenderer.default;
  }
});
Object.defineProperty(exports, "splitBoxModelStyle", {
  enumerable: true,
  get: function () {
    return _splitBoxModelStyle.default;
  }
});
Object.defineProperty(exports, "useSharedProps", {
  enumerable: true,
  get: function () {
    return _SharedPropsContext.useSharedProps;
  }
});
exports.default = void 0;

var _RenderHTML = _interopRequireDefault(require("./RenderHTML"));

var _transientRenderEngine = require("@native-html/transient-render-engine");

var _sharedTypes = require("./shared-types");

Object.keys(_sharedTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sharedTypes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sharedTypes[key];
    }
  });
});

var _renderTypes = require("./render/render-types");

Object.keys(_renderTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _renderTypes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _renderTypes[key];
    }
  });
});

var _TChildrenRenderer = _interopRequireDefault(require("./TChildrenRenderer"));

var _TNodeChildrenRenderer = _interopRequireWildcard(require("./TNodeChildrenRenderer"));

var _TNodeRenderer = _interopRequireDefault(require("./TNodeRenderer"));

var _useTRenderEngine = _interopRequireDefault(require("./hooks/useTRenderEngine"));

var _useTTree = _interopRequireDefault(require("./hooks/useTTree"));

var _defaultRenderers = _interopRequireDefault(require("./render/defaultRenderers"));

var _extendDefaultRenderer = _interopRequireDefault(require("./render/extendDefaultRenderer"));

var _splitBoxModelStyle = _interopRequireDefault(require("./helpers/splitBoxModelStyle"));

var _SharedPropsContext = require("./context/SharedPropsContext");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _RenderHTML.default;
exports.default = _default;
