// react-native-render-html/packages/render-html/lib/commonjs/RenderHTML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = RenderHTML;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactNative = require("react-native");

var _TNodeRenderer = _interopRequireDefault(require("./TNodeRenderer"));

var _useTTree = _interopRequireDefault(require("./hooks/useTTree"));

var _SharedPropsContext = _interopRequireWildcard(require("./context/SharedPropsContext"));

var _TChildrenRendererContext = _interopRequireDefault(require("./context/TChildrenRendererContext"));

var _TNodeChildrenRenderer = _interopRequireDefault(require("./TNodeChildrenRenderer"));

var _RenderHTMLDebug = _interopRequireDefault(require("./RenderHTMLDebug"));

var _LoadHTML = _interopRequireDefault(require("./LoadHTML"));

var _RenderRegistryProvider = _interopRequireDefault(require("./context/RenderRegistryProvider"));

var _TChildrenRenderer = _interopRequireDefault(require("./TChildrenRenderer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const propTypes = {
  renderers: _propTypes.default.object.isRequired,
  enableCSSInlineProcessing: _propTypes.default.bool,
  enableUserAgentStyles: _propTypes.default.bool,
  enableExperimentalMarginCollapsing: _propTypes.default.bool,
  idsStyles: _propTypes.default.object,
  remoteErrorView: _propTypes.default.func,
  remoteLoadingView: _propTypes.default.func,
  ignoredTags: _propTypes.default.array.isRequired,
  ignoredStyles: _propTypes.default.array.isRequired,
  allowedStyles: _propTypes.default.array,
  decodeEntities: _propTypes.default.bool.isRequired,
  debug: _propTypes.default.bool.isRequired,
  listsPrefixesRenderers: _propTypes.default.object,
  alterData: _propTypes.default.func,
  alterChildren: _propTypes.default.func,
  alterNode: _propTypes.default.func,
  ignoreNode: _propTypes.default.func,
  html: _propTypes.default.string,
  uri: _propTypes.default.string,
  tagsStyles: _propTypes.default.object,
  classesStyles: _propTypes.default.object,
  containerStyle: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array]),
  customWrapper: _propTypes.default.func,
  onLinkPress: _propTypes.default.func,
  computeImagesMaxWidth: _propTypes.default.func,
  staticContentMaxWidth: _propTypes.default.number,
  contentWidth: _propTypes.default.number,
  enableExperimentalPercentWidth: _propTypes.default.bool,
  imagesInitialDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number
  }),
  emSize: _propTypes.default.number.isRequired,
  ptSize: _propTypes.default.number.isRequired,
  baseStyle: _propTypes.default.object,
  textSelectable: _propTypes.default.bool,
  renderersProps: _propTypes.default.object,
  allowFontScaling: _propTypes.default.bool,
  onTTreeChange: _propTypes.default.func,
  onHTMLLoaded: _propTypes.default.func,
  systemFonts: _propTypes.default.arrayOf(_propTypes.default.string),
  fallbackFonts: _propTypes.default.shape({
    serif: _propTypes.default.string,
    'sans-serif': _propTypes.default.string,
    monospace: _propTypes.default.string
  }),
  triggerTREInvalidationPropNames: _propTypes.default.arrayOf(_propTypes.default.string)
};
const defaultProps = { ..._SharedPropsContext.defaultSharedPropsContext,
  decodeEntities: true,
  emSize: 14,
  ptSize: 1.3,
  ignoredTags: [],
  ignoredStyles: [],
  baseStyle: {
    fontSize: 14
  },
  tagsStyles: {},
  classesStyles: {},
  enableUserAgentStyles: true,
  enableCSSInlineProcessing: true,
  renderers: {},
  fallbackFonts: {
    'sans-serif': _reactNative.Platform.select({
      ios: 'system',
      default: 'sans-serif'
    }),
    monospace: _reactNative.Platform.select({
      ios: 'Menlo',
      default: 'monospace'
    }),
    serif: _reactNative.Platform.select({
      ios: 'Times New Roman',
      default: 'serif'
    })
  },
  systemFonts: _reactNative.Platform.select({
    default: [],
    ios: ['San Francisco', 'Arial', 'ArialHebrew', 'Avenir', 'Baskerville', 'Bodoni 72', 'Bradley Hand', 'Chalkboard SE', 'Cochin', 'Copperplate', 'Courier', 'Courier New', 'Damascus', 'Didot', 'Futura', 'Geeza Pro', 'Georgia', 'Gill Sans', 'Helvetica', 'Helvetica Neue', 'Hiragino Sans', 'Hoefler Text', 'Iowan Old Style', 'Kailasa', 'Khmer Sangam MN', 'Marker Felt', 'Menlo', 'Mishafi', 'Noteworthy', 'Optima', 'Palatino', 'Papyrus', 'Savoye LET', 'Symbol', 'Thonburi', 'Times New Roman', 'Trebuchet MS', 'Verdana', 'Zapf Dingbats', 'Zapfino'],
    android: ['Roboto', 'notoserif', 'sans-serif-light', 'sans-serif-thin', 'sans-serif-medium']
  }),
  triggerTREInvalidationPropNames: [],
  debug: __DEV__
};

function RenderResolvedHTML(props) {
  const ttree = (0, _useTTree.default)(props);
  return /*#__PURE__*/_react.default.createElement(_TNodeRenderer.default, {
    hasAnchorAncestor: false,
    tnode: ttree,
    collapsedMarginTop: null
  });
}

function RenderHTML(props) {
  return /*#__PURE__*/_react.default.createElement(_RenderHTMLDebug.default, props, /*#__PURE__*/_react.default.createElement(_RenderRegistryProvider.default, {
    renderers: props.renderers
  }, /*#__PURE__*/_react.default.createElement(_SharedPropsContext.default.Provider, {
    value: props
  }, /*#__PURE__*/_react.default.createElement(_TChildrenRendererContext.default.Provider, {
    value: (0, _react.useMemo)(() => ({
      TChildrenRenderer: _TChildrenRenderer.default,
      TNodeChildrenRenderer: _TNodeChildrenRenderer.default
    }), [])
  }, /*#__PURE__*/_react.default.createElement(_LoadHTML.default, props, resolvedHTML => /*#__PURE__*/_react.default.createElement(RenderResolvedHTML, _extends({}, props, {
    html: resolvedHTML
  })))))));
}

RenderHTML.defaultProps = defaultProps;
RenderHTML.propTypes = propTypes;
