// wrn-core/dist/index.js
import { StyleSheet, Dimensions, Animated, Platform, Text as Text$1, View as View$1, Easing, BackHandler, TextInput, TouchableOpacity, ActivityIndicator, Image as Image$1, Modal as Modal$1, FlatList as FlatList$1, ScrollView, Linking, Keyboard, LayoutAnimation } from 'react-native';
import React, { useEffect, useRef, useState, useMemo, forwardRef, memo, useContext, createContext, useCallback, PureComponent } from 'react';
import { useSafeArea, SafeAreaProvider } from 'react-native-safe-area-context';
import { BaseButton } from 'react-native-gesture-handler';
import { Feather, FontAwesome, FontAwesome5, MaterialIcons, MaterialCommunityIcons, AntDesign } from '@expo/vector-icons';
import { path, equals, includes, range, pipe, groupWith, type, map, isEmpty, insert } from 'ramda';
import NetInfo from '@react-native-community/netinfo';
import { WebView } from 'react-native-webview';
import { withNavigation } from 'react-navigation';
import { openBrowserAsync } from 'expo-web-browser';
import axios from 'axios';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function __spreadArray(to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
}

var defaultColors = {
  primary: '#3369e7',
  secondary: '#2dde98',
  tertiary: '#ff6c5f',
  quaternary: '#ffc20e',
  success: '#5cb85c',
  warning: '#f0ad4e',
  danger: '#fd5c63',
  info: '#5bc0de',
  facebook: '#325d94',
  twitter: '#00aadb',
  tumblr: '#2f4e6b',
  dribbble: '#fb4087',
  youtube: '#df2e1c',
  vk: '#4c75a3',
  digg: '#1b5891',
  reddit: '#ff4500',
  linkedin: '#1686b0',
  stumbleupon: '#eb4924',
  vimeo: '#63b3e4',
  instagram: '#517fa4',
  pinterest: '#cc1d24',
  behance: '#1478ff',
  heart: '#4bd1fa',
  github: '#24292e',
  dark1: '#111111',
  dark2: '#444444',
  dark3: '#888888',
  dark4: '#cccccc',
  gray1: '#d6d6da',
  gray2: '#e7e7ed',
  gray3: '#f0f0f3',
  gray4: '#fafafb',
  gray5: '#fbfbfc',
  light: '#fff',
  transparent: 'transparent' };

var defaultSizes = {
  base: 15,
  font: 14,
  container: 600 };

var getDefaultStyles = function (colors) {
  return StyleSheet.create({
    colorPrimary: {
      color: colors.primary },

    colorSecondary: {
      color: colors.secondary },

    colorTertiary: {
      color: colors.tertiary },

    colorQuaternary: {
      color: colors.quaternary },

    colorInfo: {
      color: colors.info },

    colorSuccess: {
      color: colors.success },

    colorWarning: {
      color: colors.warning },

    colorDanger: {
      color: colors.danger },

    colorDark1: {
      color: colors.dark1 },

    colorDark2: {
      color: colors.dark2 },

    colorDark3: {
      color: colors.dark3 },

    colorDark4: {
      color: colors.dark4 },

    colorGray1: {
      color: colors.gray1 },

    colorGray2: {
      color: colors.gray2 },

    colorGray3: {
      color: colors.gray3 },

    colorGray4: {
      color: colors.gray4 },

    colorLight: {
      color: colors.light },

    bgDark1: {
      backgroundColor: colors.dark1 },

    bgDark2: {
      backgroundColor: colors.dark2 },

    bgDark3: {
      backgroundColor: colors.dark3 },

    bgDark4: {
      backgroundColor: colors.dark4 },

    bgGray1: {
      backgroundColor: colors.gray1 },

    bgGray2: {
      backgroundColor: colors.gray2 },

    bgGray3: {
      backgroundColor: colors.gray3 },

    bgGray4: {
      backgroundColor: colors.gray4 },

    bgGray5: {
      backgroundColor: colors.gray5 },

    bgLight: {
      backgroundColor: colors.light },

    bgPrimary: {
      backgroundColor: colors.primary },

    bgTransparent: {
      backgroundColor: 'transparent' },

    bgInfo: {
      backgroundColor: colors.info },

    bgSuccess: {
      backgroundColor: colors.success },

    bgWarning: {
      backgroundColor: colors.warning },

    bgDanger: {
      backgroundColor: colors.danger },

    bDark1: {
      borderColor: colors.dark1 },

    bDark2: {
      borderColor: colors.dark2 },

    bDark3: {
      borderColor: colors.dark3 },

    bDark4: {
      borderColor: colors.dark4 },

    bGray1: {
      borderColor: colors.gray1 },

    bGray2: {
      borderColor: colors.gray2 },

    bGray3: {
      borderColor: colors.gray3 },

    bGray4: {
      borderColor: colors.gray4 },

    bGray5: {
      borderColor: colors.gray5 },

    bLight: {
      borderColor: colors.light },

    bPrimary: {
      borderColor: colors.primary },

    bTransparent: {
      borderColor: 'transparent' },

    bInfo: {
      borderColor: colors.info },

    bSuccess: {
      borderColor: colors.success },

    bWarning: {
      borderColor: colors.warning },

    bDanger: {
      borderColor: colors.danger } });


};

var Emitter = /** @class */function () {
  function Emitter() {
    this.events = {};
    this.id = 1;
  }
  Emitter.prototype.emit = function (type, payload) {
    if (this.events[type]) {
      this.events[type].forEach(function (_a) {
        var listener = _a.listener;
        listener(payload);
      });
    }
  };
  Emitter.prototype.on = function (type, listener) {
    var _a;
    var id = ++this.id;
    this.events = __assign(__assign({}, this.events), (_a = {}, _a[type] = __spreadArray(__spreadArray([], this.events[type] || []), [
    {
      listener: listener,
      id: id }]),

    _a));
    return id;
  };
  Emitter.prototype.once = function (type, listener) {
    var _a;
    var id = ++this.id;
    this.events = __assign(__assign({}, this.events), (_a = {}, _a[type] = [{ listener: listener, id: id }], _a));
    return id;
  };
  Emitter.prototype.off = function (id) {
    var _this = this;
    this.events = Object.keys(this.events).reduce(function (obj, key) {
      var _a;
      return __assign(__assign({}, obj), (_a = {}, _a[key] = _this.events[key].filter(function (item) {return item.id !== id;}), _a));
    }, {});
  };
  Emitter.prototype.offReference = function (type, listener) {
    if (this.events[type]) {
      this.events[type] = this.events[type].filter(function (item) {
        return item.listener !== listener;
      });
    }
  };
  Emitter.prototype.offEvent = function (type) {
    var _this = this;
    if (this.events[type]) {
      this.events = Object.keys(this.events).reduce(function (obj, key) {
        var _a;
        return __assign(__assign({}, obj), key === type ? {} : (_a = {}, _a[key] = _this.events[key], _a));
      }, {});
    }
  };
  return Emitter;
}();

function useMount(effect) {
  useEffect(effect, []);
}

var useUnmount = function (fn) {
  var fnRef = useRef(fn);
  // update the ref each render so if it change the newest callback will be invoked
  fnRef.current = fn;
  useMount(function () {return function () {return fnRef.current();};});
};

var SCREEN_WIDTH = Dimensions.get('window').width;
var SCREEN_HEIGHT = Dimensions.get('window').height;

function usePortalAnimation(_a) {
  var visible = _a.visible,animationType = _a.animationType,animationDuration = _a.animationDuration,animmationEasing = _a.animmationEasing,_b = _a.elementHeight,elementHeight = _b === void 0 ? SCREEN_HEIGHT : _b,_c = _a.elementWidth,elementWidth = _c === void 0 ? SCREEN_WIDTH : _c;
  var _d = useState(visible),visibleState = _d[0],setVisibleState = _d[1];
  var animmationEasingRef = useRef(animmationEasing);
  var animation = useRef(new Animated.Value(0));
  useEffect(function () {
    visible && setVisibleState(true);
    Animated.timing(animation.current, {
      toValue: visible ? 100 : 0,
      duration: animationType === 'none' ? 0 : animationDuration,
      easing: animmationEasingRef.current,
      useNativeDriver: true }).
    start(function () {
      setVisibleState(visible);
    });
  }, [animationDuration, animationType, visible]);
  var slide = useMemo(function () {
    var checkNegativeAndPositiveNumber = animationType === 'slide-up' || animationType === 'slide-left';
    var checkVertical = animationType === 'slide-up' || animationType === 'slide-down';
    var switchNegativeAndPositiveNumber = checkNegativeAndPositiveNumber ? 1 : -1;
    var outputFirst = (checkVertical ? elementHeight : elementWidth) * switchNegativeAndPositiveNumber;
    return animation.current.interpolate({
      inputRange: [0, 100],
      outputRange: [outputFirst, 0],
      extrapolate: 'clamp' });

  }, [animationType, elementHeight, elementWidth]);
  var fade = animation.current.interpolate({
    inputRange: [0, 100],
    outputRange: [0, 1],
    extrapolate: 'clamp' });

  var zoom = animation.current.interpolate({
    inputRange: [0, 100],
    outputRange: [0.4, 1],
    extrapolate: 'clamp' });

  var checkAnimated = function () {
    switch (animationType) {
      case 'slide-up':
      case 'slide-down':
        return { transform: [{ translateY: slide }] };
      case 'slide-left':
      case 'slide-right':
        return { transform: [{ translateX: slide }] };
      case 'zoom':
        return { transform: [{ scale: zoom }], opacity: fade };
      case 'fade':
        return { opacity: fade };
      default:
        return {};}

  };
  return { visibleState: visibleState, animationStyle: checkAnimated() };
}

var tachyonsStyles = StyleSheet.create({
  flex: { flex: 1 },
  flex0: { flex: 0 },
  flex1: { flex: 0.1 },
  flex2: { flex: 0.2 },
  flex3: { flex: 0.3 },
  flex4: { flex: 0.4 },
  flex5: { flex: 0.5 },
  flex6: { flex: 0.6 },
  flex7: { flex: 0.7 },
  flex8: { flex: 0.8 },
  flex9: { flex: 0.9 },
  noFlex: { flex: -1 },
  contain: { resizeMode: 'contain' },
  cover: { resizeMode: 'cover' },
  stretch: { resizeMode: 'stretch' },
  ba: { borderStyle: 'solid', borderWidth: 1 },
  bt: { borderTopWidth: 1 },
  br: { borderRightWidth: 1 },
  bb: { borderBottomWidth: 1 },
  bl: { borderLeftWidth: 1 },
  bn: { borderWidth: 0 },
  bBlack: { borderColor: '#000' },
  bNearBlack: { borderColor: '#111' },
  bDarkGray: { borderColor: '#333' },
  bMidGray: { borderColor: '#555' },
  bGray: { borderColor: '#777' },
  bSilver: { borderColor: '#999' },
  bLightSilver: { borderColor: '#aaa' },
  bMoonGray: { borderColor: '#ccc' },
  bLightGray: { borderColor: '#eee' },
  bNearWhite: { borderColor: '#f4f4f4' },
  bWhite: { borderColor: '#fff' },
  bWhite90: { borderColor: 'rgba( 255, 255, 255, .9 )' },
  bWhite80: { borderColor: 'rgba( 255, 255, 255, .8 )' },
  bWhite70: { borderColor: 'rgba( 255, 255, 255, .7 )' },
  bWhite60: { borderColor: 'rgba( 255, 255, 255, .6 )' },
  bWhite50: { borderColor: 'rgba( 255, 255, 255, .5 )' },
  bWhite40: { borderColor: 'rgba( 255, 255, 255, .4 )' },
  bWhite30: { borderColor: 'rgba( 255, 255, 255, .3 )' },
  bWhite20: { borderColor: 'rgba( 255, 255, 255, .2 )' },
  bWhite10: { borderColor: 'rgba( 255, 255, 255, .1 )' },
  bWhite05: { borderColor: 'rgba( 255, 255, 255, .05 )' },
  bWhite025: { borderColor: 'rgba( 255, 255, 255, .025 )' },
  bWhite0125: { borderColor: 'rgba( 255, 255, 255, .0125 )' },
  bBlack90: { borderColor: 'rgba( 0, 0, 0, .9 )' },
  bBlack80: { borderColor: 'rgba( 0, 0, 0, .8 )' },
  bBlack70: { borderColor: 'rgba( 0, 0, 0, .7 )' },
  bBlack60: { borderColor: 'rgba( 0, 0, 0, .6 )' },
  bBlack50: { borderColor: 'rgba( 0, 0, 0, .5 )' },
  bBlack40: { borderColor: 'rgba( 0, 0, 0, .4 )' },
  bBlack30: { borderColor: 'rgba( 0, 0, 0, .3 )' },
  bBlack20: { borderColor: 'rgba( 0, 0, 0, .2 )' },
  bBlack10: { borderColor: 'rgba( 0, 0, 0, .1 )' },
  bBlack05: { borderColor: 'rgba( 0, 0, 0, .05 )' },
  bBlack025: { borderColor: 'rgba( 0, 0, 0, .025 )' },
  bBlack0125: { borderColor: 'rgba( 0, 0, 0, .0125 )' },
  bDarkRed: { borderColor: '#e7040f' },
  bRed: { borderColor: '#ff4136' },
  bLightRed: { borderColor: '#ff725c' },
  bOrange: { borderColor: '#ff6300' },
  bGold: { borderColor: '#ffb700' },
  bYellow: { borderColor: '#ffd700' },
  bLightYellow: { borderColor: '#fbf1a9' },
  bPurple: { borderColor: '#5e2ca5' },
  bLightPurple: { borderColor: '#a463f2' },
  bDarkPink: { borderColor: '#d5008f' },
  bHotPink: { borderColor: '#ff41b4' },
  bPink: { borderColor: '#ff80cc' },
  bLightPink: { borderColor: '#ffa3d7' },
  bDarkGreen: { borderColor: '#137752' },
  bGreen: { borderColor: '#19a974' },
  bLightGreen: { borderColor: '#9eebcf' },
  bNavy: { borderColor: '#001b44' },
  bDarkBlue: { borderColor: '#00449e' },
  bBlue: { borderColor: '#357edd' },
  bLightBlue: { borderColor: '#96ccff' },
  bLightestBlue: { borderColor: '#cdecff' },
  bWashedBlue: { borderColor: '#f6fffe' },
  bWashedGreen: { borderColor: '#e8fdf5' },
  bWashedYellow: { borderColor: '#fffceb' },
  bWashedRed: { borderColor: '#ffdfdf' },
  bTransparent: { borderColor: 'transparent' },
  br0: { borderRightWidth: 0 },
  br1: { borderRadius: 2 },
  br2: { borderRadius: 4 },
  br3: { borderRadius: 8 },
  br4: { borderRadius: 16 },
  brPill: { borderRadius: 9999 },
  brBottom: { borderTopLeftRadius: 0, borderTopRightRadius: 0 },
  brTop: { borderBottomLeftRadius: 0, borderBottomRightRadius: 0 },
  brRight: { borderTopLeftRadius: 0, borderBottomLeftRadius: 0 },
  brLeft: { borderTopRightRadius: 0, borderBottomRightRadius: 0 },
  bDotted: { borderStyle: 'dotted' },
  bDashed: { borderStyle: 'dashed' },
  bSolid: { borderStyle: 'solid' },
  bw0: { borderWidth: 0 },
  bw1: { borderWidth: 2 },
  bw2: { borderWidth: 4 },
  bw3: { borderWidth: 8 },
  bw4: { borderWidth: 16 },
  bw5: { borderWidth: 32 },
  bt0: { borderTopWidth: 0 },
  bb0: { borderBottomWidth: 0 },
  bl0: { borderLeftWidth: 0 },
  top0: { top: 0 },
  right0: { right: 0 },
  bottom0: { bottom: 0 },
  left0: { left: 0 },
  top1: { top: 16 },
  right1: { right: 16 },
  bottom1: { bottom: 16 },
  left1: { left: 16 },
  top2: { top: 32 },
  right2: { right: 32 },
  bottom2: { bottom: 32 },
  left2: { left: 32 },
  absoluteFill: { top: 0, right: 0, bottom: 0, left: 0 },
  dtFixed: { width: '100%' },
  flexAuto: { minWidth: 0 },
  flexColumn: { flexDirection: 'column' },
  flexRow: { flexDirection: 'row' },
  flexWrap: { flexWrap: 'wrap' },
  flexWrapReverse: { flexWrap: 'wrap-reverse' },
  flexNowrap: { flexWrap: 'nowrap' },
  flexColumnReverse: { flexDirection: 'column-reverse' },
  flexRowReverse: { flexDirection: 'row-reverse' },
  itemsStart: { alignItems: 'flex-start' },
  itemsEnd: { alignItems: 'flex-end' },
  itemsCenter: { alignItems: 'center' },
  itemsBaseline: { alignItems: 'baseline' },
  itemsStretch: { alignItems: 'stretch' },
  selfStart: { alignSelf: 'flex-start' },
  selfEnd: { alignSelf: 'flex-end' },
  selfCenter: { alignSelf: 'center' },
  selfBaseline: { alignSelf: 'baseline' },
  selfStretch: { alignSelf: 'stretch' },
  justifyStart: { justifyContent: 'flex-start' },
  justifyEnd: { justifyContent: 'flex-end' },
  justifyCenter: { justifyContent: 'center' },
  justifyBetween: { justifyContent: 'space-between' },
  justifyAround: { justifyContent: 'space-around' },
  justifyEvenly: { justifyContent: 'space-evenly' },
  i: { fontStyle: 'italic' },
  fsNormal: { fontStyle: 'normal' },
  normal: { fontWeight: 'normal' },
  b: { fontWeight: 'bold' },
  fw1: { fontWeight: '100' },
  fw2: { fontWeight: '200' },
  fw3: { fontWeight: '300' },
  fw4: { fontWeight: '400' },
  fw5: { fontWeight: '500' },
  fw6: { fontWeight: '600' },
  fw7: { fontWeight: '700' },
  fw8: { fontWeight: '800' },
  fw9: { fontWeight: '900' },
  h1: { height: 16 },
  h2: { height: 32 },
  h3: { height: 64 },
  h4: { height: 128 },
  h5: { height: 256 },
  h25: { height: '25%' },
  h50: { height: '50%' },
  h75: { height: '75%' },
  h100: { height: '100%' },
  minH100: { minHeight: '100%' },
  hAuto: { height: 'auto' },
  tracked: { letterSpacing: 1.6 },
  trackedTight: { letterSpacing: -0.8 },
  trackedMega: { letterSpacing: 4 },
  lhSolid: { lineHeight: 1 },
  lhTitle: { lineHeight: 1.25 },
  lhCopy: { lineHeight: 1.5 },
  mw100: { maxWidth: '100%' },
  mw1: { maxWidth: 16 },
  mw2: { maxWidth: 32 },
  mw3: { maxWidth: 64 },
  mw4: { maxWidth: 128 },
  mw5: { maxWidth: 256 },
  mw6: { maxWidth: 512 },
  mw7: { maxWidth: 768 },
  mw8: { maxWidth: 1024 },
  mw9: { maxWidth: 1536 },
  mwNone: { maxWidth: 'none' },
  w1: { width: 16 },
  w2: { width: 32 },
  w3: { width: 64 },
  w4: { width: 128 },
  w5: { width: 256 },
  w10: { width: '10%' },
  w20: { width: '20%' },
  w25: { width: '25%' },
  w30: { width: '30%' },
  w33: { width: '33%' },
  w34: { width: '34%' },
  w40: { width: '40%' },
  w50: { width: '50%' },
  w60: { width: '60%' },
  w70: { width: '70%' },
  w75: { width: '75%' },
  w80: { width: '80%' },
  w90: { width: '90%' },
  w100: { width: '100%' },
  wAuto: { width: 'auto' },
  overflowVisible: { overflow: 'visible' },
  overflowHidden: { overflow: 'hidden' },
  relative: { position: 'relative' },
  absolute: { position: 'absolute' },
  o100: { opacity: 1 },
  o90: { opacity: 0.9 },
  o80: { opacity: 0.8 },
  o70: { opacity: 0.7 },
  o60: { opacity: 0.6 },
  o50: { opacity: 0.5 },
  o40: { opacity: 0.4 },
  o30: { opacity: 0.3 },
  o20: { opacity: 0.2 },
  o10: { opacity: 0.1 },
  o05: { opacity: 0.05 },
  o025: { opacity: 0.025 },
  o0: { opacity: 0 },
  black90: { color: 'rgba( 0, 0, 0, .9 )' },
  black80: { color: 'rgba( 0, 0, 0, .8 )' },
  black70: { color: 'rgba( 0, 0, 0, .7 )' },
  black60: { color: 'rgba( 0, 0, 0, .6 )' },
  black50: { color: 'rgba( 0, 0, 0, .5 )' },
  black40: { color: 'rgba( 0, 0, 0, .4 )' },
  black30: { color: 'rgba( 0, 0, 0, .3 )' },
  black20: { color: 'rgba( 0, 0, 0, .2 )' },
  black10: { color: 'rgba( 0, 0, 0, .1 )' },
  black05: { color: 'rgba( 0, 0, 0, .05 )' },
  white90: { color: 'rgba( 255, 255, 255, .9 )' },
  white80: { color: 'rgba( 255, 255, 255, .8 )' },
  white70: { color: 'rgba( 255, 255, 255, .7 )' },
  white60: { color: 'rgba( 255, 255, 255, .6 )' },
  white50: { color: 'rgba( 255, 255, 255, .5 )' },
  white40: { color: 'rgba( 255, 255, 255, .4 )' },
  white30: { color: 'rgba( 255, 255, 255, .3 )' },
  white20: { color: 'rgba( 255, 255, 255, .2 )' },
  white10: { color: 'rgba( 255, 255, 255, .1 )' },
  black: { color: '#000' },
  nearBlack: { color: '#111' },
  darkGray: { color: '#333' },
  midGray: { color: '#555' },
  gray: { color: '#777' },
  silver: { color: '#999' },
  lightSilver: { color: '#aaa' },
  moonGray: { color: '#ccc' },
  lightGray: { color: '#eee' },
  nearWhite: { color: '#f4f4f4' },
  white: { color: '#fff' },
  darkRed: { color: '#e7040f' },
  red: { color: '#ff4136' },
  lightRed: { color: '#ff725c' },
  orange: { color: '#ff6300' },
  gold: { color: '#ffb700' },
  yellow: { color: '#ffd700' },
  lightYellow: { color: '#fbf1a9' },
  purple: { color: '#5e2ca5' },
  lightPurple: { color: '#a463f2' },
  darkPink: { color: '#d5008f' },
  hotPink: { color: '#ff41b4' },
  pink: { color: '#ff80cc' },
  lightPink: { color: '#ffa3d7' },
  darkGreen: { color: '#137752' },
  green: { color: '#19a974' },
  lightGreen: { color: '#9eebcf' },
  navy: { color: '#001b44' },
  darkBlue: { color: '#00449e' },
  blue: { color: '#357edd' },
  lightBlue: { color: '#96ccff' },
  lightestBlue: { color: '#cdecff' },
  washedBlue: { color: '#f6fffe' },
  washedGreen: { color: '#e8fdf5' },
  washedYellow: { color: '#fffceb' },
  washedRed: { color: '#ffdfdf' },
  bgBlack90: { backgroundColor: 'rgba( 0, 0, 0, .9 )' },
  bgBlack80: { backgroundColor: 'rgba( 0, 0, 0, .8 )' },
  bgBlack70: { backgroundColor: 'rgba( 0, 0, 0, .7 )' },
  bgBlack60: { backgroundColor: 'rgba( 0, 0, 0, .6 )' },
  bgBlack50: { backgroundColor: 'rgba( 0, 0, 0, .5 )' },
  bgBlack40: { backgroundColor: 'rgba( 0, 0, 0, .4 )' },
  bgBlack30: { backgroundColor: 'rgba( 0, 0, 0, .3 )' },
  bgBlack20: { backgroundColor: 'rgba( 0, 0, 0, .2 )' },
  bgBlack10: { backgroundColor: 'rgba( 0, 0, 0, .1 )' },
  bgBlack05: { backgroundColor: 'rgba( 0, 0, 0, .05 )' },
  bgWhite90: { backgroundColor: 'rgba( 255, 255, 255, .9 )' },
  bgWhite80: { backgroundColor: 'rgba( 255, 255, 255, .8 )' },
  bgWhite70: { backgroundColor: 'rgba( 255, 255, 255, .7 )' },
  bgWhite60: { backgroundColor: 'rgba( 255, 255, 255, .6 )' },
  bgWhite50: { backgroundColor: 'rgba( 255, 255, 255, .5 )' },
  bgWhite40: { backgroundColor: 'rgba( 255, 255, 255, .4 )' },
  bgWhite30: { backgroundColor: 'rgba( 255, 255, 255, .3 )' },
  bgWhite20: { backgroundColor: 'rgba( 255, 255, 255, .2 )' },
  bgWhite10: { backgroundColor: 'rgba( 255, 255, 255, .1 )' },
  bgBlack: { backgroundColor: '#000' },
  bgNearBlack: { backgroundColor: '#111' },
  bgDarkGray: { backgroundColor: '#333' },
  bgMidGray: { backgroundColor: '#555' },
  bgGray: { backgroundColor: '#777' },
  bgSilver: { backgroundColor: '#999' },
  bgLightSilver: { backgroundColor: '#aaa' },
  bgMoonGray: { backgroundColor: '#ccc' },
  bgLightGray: { backgroundColor: '#eee' },
  bgNearWhite: { backgroundColor: '#f4f4f4' },
  bgWhite: { backgroundColor: '#fff' },
  bgTransparent: { backgroundColor: 'transparent' },
  bgDarkRed: { backgroundColor: '#e7040f' },
  bgRed: { backgroundColor: '#ff4136' },
  bgLightRed: { backgroundColor: '#ff725c' },
  bgOrange: { backgroundColor: '#ff6300' },
  bgGold: { backgroundColor: '#ffb700' },
  bgYellow: { backgroundColor: '#ffd700' },
  bgLightYellow: { backgroundColor: '#fbf1a9' },
  bgPurple: { backgroundColor: '#5e2ca5' },
  bgLightPurple: { backgroundColor: '#a463f2' },
  bgDarkPink: { backgroundColor: '#d5008f' },
  bgHotPink: { backgroundColor: '#ff41b4' },
  bgPink: { backgroundColor: '#ff80cc' },
  bgLightPink: { backgroundColor: '#ffa3d7' },
  bgDarkGreen: { backgroundColor: '#137752' },
  bgGreen: { backgroundColor: '#19a974' },
  bgLightGreen: { backgroundColor: '#9eebcf' },
  bgNavy: { backgroundColor: '#001b44' },
  bgDarkBlue: { backgroundColor: '#00449e' },
  bgBlue: { backgroundColor: '#357edd' },
  bgLightBlue: { backgroundColor: '#96ccff' },
  bgLightestBlue: { backgroundColor: '#cdecff' },
  bgWashedBlue: { backgroundColor: '#f6fffe' },
  bgWashedGreen: { backgroundColor: '#e8fdf5' },
  bgWashedYellow: { backgroundColor: '#fffceb' },
  bgWashedRed: { backgroundColor: '#ffdfdf' },
  pa0: { padding: 0 },
  pa1: { padding: 4 },
  pa2: { padding: 8 },
  pa3: { padding: 16 },
  pa4: { padding: 32 },
  pa5: { padding: 64 },
  pa6: { padding: 128 },
  pa7: { padding: 256 },
  pl0: { paddingLeft: 0 },
  pl1: { paddingLeft: 4 },
  pl2: { paddingLeft: 8 },
  pl3: { paddingLeft: 16 },
  pl4: { paddingLeft: 32 },
  pl5: { paddingLeft: 64 },
  pl6: { paddingLeft: 128 },
  pl7: { paddingLeft: 256 },
  pr0: { paddingRight: 0 },
  pr1: { paddingRight: 4 },
  pr2: { paddingRight: 8 },
  pr3: { paddingRight: 16 },
  pr4: { paddingRight: 32 },
  pr5: { paddingRight: 64 },
  pr6: { paddingRight: 128 },
  pr7: { paddingRight: 256 },
  pb0: { paddingBottom: 0 },
  pb1: { paddingBottom: 4 },
  pb2: { paddingBottom: 8 },
  pb3: { paddingBottom: 16 },
  pb4: { paddingBottom: 32 },
  pb5: { paddingBottom: 64 },
  pb6: { paddingBottom: 128 },
  pb7: { paddingBottom: 256 },
  pt0: { paddingTop: 0 },
  pt1: { paddingTop: 4 },
  pt2: { paddingTop: 8 },
  pt3: { paddingTop: 16 },
  pt4: { paddingTop: 32 },
  pt5: { paddingTop: 64 },
  pt6: { paddingTop: 128 },
  pt7: { paddingTop: 256 },
  pv0: { paddingVertical: 0 },
  pv1: { paddingVertical: 4 },
  pv2: { paddingVertical: 8 },
  pv3: { paddingVertical: 16 },
  pv4: { paddingVertical: 32 },
  pv5: { paddingVertical: 64 },
  pv6: { paddingVertical: 128 },
  pv7: { paddingVertical: 256 },
  ph0: { paddingHorizontal: 0 },
  ph1: { paddingHorizontal: 4 },
  ph2: { paddingHorizontal: 8 },
  ph3: { paddingHorizontal: 16 },
  ph4: { paddingHorizontal: 32 },
  ph5: { paddingHorizontal: 64 },
  ph6: { paddingHorizontal: 128 },
  ph7: { paddingHorizontal: 256 },
  ma0: { margin: 0 },
  ma1: { margin: 4 },
  ma2: { margin: 8 },
  ma3: { margin: 16 },
  ma4: { margin: 32 },
  ma5: { margin: 64 },
  ma6: { margin: 128 },
  ma7: { margin: 256 },
  ml0: { marginLeft: 0 },
  ml1: { marginLeft: 4 },
  ml2: { marginLeft: 8 },
  ml3: { marginLeft: 16 },
  ml4: { marginLeft: 32 },
  ml5: { marginLeft: 64 },
  ml6: { marginLeft: 128 },
  ml7: { marginLeft: 256 },
  mr0: { marginRight: 0 },
  mr1: { marginRight: 4 },
  mr2: { marginRight: 8 },
  mr3: { marginRight: 16 },
  mr4: { marginRight: 32 },
  mr5: { marginRight: 64 },
  mr6: { marginRight: 128 },
  mr7: { marginRight: 256 },
  mb0: { marginBottom: 0 },
  mb1: { marginBottom: 4 },
  mb2: { marginBottom: 8 },
  mb3: { marginBottom: 16 },
  mb4: { marginBottom: 32 },
  mb5: { marginBottom: 64 },
  mb6: { marginBottom: 128 },
  mb7: { marginBottom: 256 },
  mt0: { marginTop: 0 },
  mt1: { marginTop: 4 },
  mt2: { marginTop: 8 },
  mt3: { marginTop: 16 },
  mt4: { marginTop: 32 },
  mt5: { marginTop: 64 },
  mt6: { marginTop: 128 },
  mt7: { marginTop: 256 },
  mv0: { marginVertical: 0 },
  mv1: { marginVertical: 4 },
  mv2: { marginVertical: 8 },
  mv3: { marginVertical: 16 },
  mv4: { marginVertical: 32 },
  mv5: { marginVertical: 64 },
  mv6: { marginVertical: 128 },
  mv7: { marginVertical: 256 },
  mh0: { marginHorizontal: 0 },
  mh1: { marginHorizontal: 4 },
  mh2: { marginHorizontal: 8 },
  mh3: { marginHorizontal: 16 },
  mh4: { marginHorizontal: 32 },
  mh5: { marginHorizontal: 64 },
  mh6: { marginHorizontal: 128 },
  mh7: { marginHorizontal: 256 },
  na1: { margin: -4 },
  na2: { margin: -8 },
  na3: { margin: -16 },
  na4: { margin: -32 },
  na5: { margin: -64 },
  na6: { margin: -128 },
  na7: { margin: -256 },
  nl1: { marginLeft: -4 },
  nl2: { marginLeft: -8 },
  nl3: { marginLeft: -16 },
  nl4: { marginLeft: -32 },
  nl5: { marginLeft: -64 },
  nl6: { marginLeft: -128 },
  nl7: { marginLeft: -256 },
  nr1: { marginRight: -4 },
  nr2: { marginRight: -8 },
  nr3: { marginRight: -16 },
  nr4: { marginRight: -32 },
  nr5: { marginRight: -64 },
  nr6: { marginRight: -128 },
  nr7: { marginRight: -256 },
  nb1: { marginBottom: -4 },
  nb2: { marginBottom: -8 },
  nb3: { marginBottom: -16 },
  nb4: { marginBottom: -32 },
  nb5: { marginBottom: -64 },
  nb6: { marginBottom: -128 },
  nb7: { marginBottom: -256 },
  nt1: { marginTop: -4 },
  nt2: { marginTop: -8 },
  nt3: { marginTop: -16 },
  nt4: { marginTop: -32 },
  nt5: { marginTop: -64 },
  nt6: { marginTop: -128 },
  nt7: { marginTop: -256 },
  tl: { textAlign: 'left' },
  tr: { textAlign: 'right' },
  tc: { textAlign: 'center' },
  f1: { fontSize: 48 },
  f2: { fontSize: 36 },
  f3: { fontSize: 24 },
  f4: { fontSize: 20 },
  f5: { fontSize: 16 },
  f6: { fontSize: 14 },
  f7: { fontSize: 12 },
  measure: { maxWidth: 480 },
  measureWide: { maxWidth: 544 },
  measureNarrow: { maxWidth: 320 },
  indent: { marginVertical: 0 },
  truncate: { overflow: 'hidden' },
  center: { marginHorizontal: 'auto' },
  mrAuto: { marginRight: 'auto' },
  mlAuto: { marginLeft: 'auto' },
  dim: { opacity: 1 },
  dimActive: { opacity: 0.8 },
  z0: { zIndex: 0 },
  z1: { zIndex: 1 },
  z2: { zIndex: 2 },
  z3: { zIndex: 3 },
  z4: { zIndex: 4 },
  z5: { zIndex: 5 },
  z999: { zIndex: 999 },
  z9999: { zIndex: 9999 },
  zMax: { zIndex: 2147483647 },
  empty: {} });

var tailwindStyles = StyleSheet.create({
  'bg-transparent': {
    backgroundColor: 'transparent' },

  'bg-black': {
    backgroundColor: '#22292f' },

  'bg-grey-darkest': {
    backgroundColor: '#3d4852' },

  'bg-grey-darker': {
    backgroundColor: '#606f7b' },

  'bg-grey-dark': {
    backgroundColor: '#8795a1' },

  'bg-grey': {
    backgroundColor: '#b8c2cc' },

  'bg-grey-light': {
    backgroundColor: '#dae1e7' },

  'bg-grey-lighter': {
    backgroundColor: '#f1f5f8' },

  'bg-grey-lightest': {
    backgroundColor: '#f8fafc' },

  'bg-white': {
    backgroundColor: '#fff' },

  'bg-red-darkest': {
    backgroundColor: '#3b0d0c' },

  'bg-red-darker': {
    backgroundColor: '#621b18' },

  'bg-red-dark': {
    backgroundColor: '#cc1f1a' },

  'bg-red': {
    backgroundColor: '#e3342f' },

  'bg-red-light': {
    backgroundColor: '#ef5753' },

  'bg-red-lighter': {
    backgroundColor: '#f9acaa' },

  'bg-red-lightest': {
    backgroundColor: '#fcebea' },

  'bg-orange-darkest': {
    backgroundColor: '#462a16' },

  'bg-orange-darker': {
    backgroundColor: '#613b1f' },

  'bg-orange-dark': {
    backgroundColor: '#de751f' },

  'bg-orange': {
    backgroundColor: '#f6993f' },

  'bg-orange-light': {
    backgroundColor: '#faad63' },

  'bg-orange-lighter': {
    backgroundColor: '#fcd9b6' },

  'bg-orange-lightest': {
    backgroundColor: '#fff5eb' },

  'bg-yellow-darkest': {
    backgroundColor: '#453411' },

  'bg-yellow-darker': {
    backgroundColor: '#684f1d' },

  'bg-yellow-dark': {
    backgroundColor: '#f2d024' },

  'bg-yellow': {
    backgroundColor: '#ffed4a' },

  'bg-yellow-light': {
    backgroundColor: '#fff382' },

  'bg-yellow-lighter': {
    backgroundColor: '#fff9c2' },

  'bg-yellow-lightest': {
    backgroundColor: '#fcfbeb' },

  'bg-green-darkest': {
    backgroundColor: '#0f2f21' },

  'bg-green-darker': {
    backgroundColor: '#1a4731' },

  'bg-green-dark': {
    backgroundColor: '#1f9d55' },

  'bg-green': {
    backgroundColor: '#38c172' },

  'bg-green-light': {
    backgroundColor: '#51d88a' },

  'bg-green-lighter': {
    backgroundColor: '#a2f5bf' },

  'bg-green-lightest': {
    backgroundColor: '#e3fcec' },

  'bg-teal-darkest': {
    backgroundColor: '#0d3331' },

  'bg-teal-darker': {
    backgroundColor: '#20504f' },

  'bg-teal-dark': {
    backgroundColor: '#38a89d' },

  'bg-teal': {
    backgroundColor: '#4dc0b5' },

  'bg-teal-light': {
    backgroundColor: '#64d5ca' },

  'bg-teal-lighter': {
    backgroundColor: '#a0f0ed' },

  'bg-teal-lightest': {
    backgroundColor: '#e8fffe' },

  'bg-blue-darkest': {
    backgroundColor: '#12283a' },

  'bg-blue-darker': {
    backgroundColor: '#1c3d5a' },

  'bg-blue-dark': {
    backgroundColor: '#2779bd' },

  'bg-blue': {
    backgroundColor: '#3490dc' },

  'bg-blue-light': {
    backgroundColor: '#6cb2eb' },

  'bg-blue-lighter': {
    backgroundColor: '#bcdefa' },

  'bg-blue-lightest': {
    backgroundColor: '#eff8ff' },

  'bg-indigo-darkest': {
    backgroundColor: '#191e38' },

  'bg-indigo-darker': {
    backgroundColor: '#2f365f' },

  'bg-indigo-dark': {
    backgroundColor: '#5661b3' },

  'bg-indigo': {
    backgroundColor: '#6574cd' },

  'bg-indigo-light': {
    backgroundColor: '#7886d7' },

  'bg-indigo-lighter': {
    backgroundColor: '#b2b7ff' },

  'bg-indigo-lightest': {
    backgroundColor: '#e6e8ff' },

  'bg-purple-darkest': {
    backgroundColor: '#21183c' },

  'bg-purple-darker': {
    backgroundColor: '#382b5f' },

  'bg-purple-dark': {
    backgroundColor: '#794acf' },

  'bg-purple': {
    backgroundColor: '#9561e2' },

  'bg-purple-light': {
    backgroundColor: '#a779e9' },

  'bg-purple-lighter': {
    backgroundColor: '#d6bbfc' },

  'bg-purple-lightest': {
    backgroundColor: '#f3ebff' },

  'bg-pink-darkest': {
    backgroundColor: '#451225' },

  'bg-pink-darker': {
    backgroundColor: '#6f213f' },

  'bg-pink-dark': {
    backgroundColor: '#eb5286' },

  'bg-pink': {
    backgroundColor: '#f66d9b' },

  'bg-pink-light': {
    backgroundColor: '#fa7ea8' },

  'bg-pink-lighter': {
    backgroundColor: '#ffbbca' },

  'bg-pink-lightest': {
    backgroundColor: '#ffebef' },

  'border-transparent': {
    borderColor: 'transparent' },

  'border-black': {
    borderColor: '#22292f' },

  'border-grey-darkest': {
    borderColor: '#3d4852' },

  'border-grey-darker': {
    borderColor: '#606f7b' },

  'border-grey-dark': {
    borderColor: '#8795a1' },

  'border-grey': {
    borderColor: '#b8c2cc' },

  'border-grey-light': {
    borderColor: '#dae1e7' },

  'border-grey-lighter': {
    borderColor: '#f1f5f8' },

  'border-grey-lightest': {
    borderColor: '#f8fafc' },

  'border-white': {
    borderColor: '#fff' },

  'border-red-darkest': {
    borderColor: '#3b0d0c' },

  'border-red-darker': {
    borderColor: '#621b18' },

  'border-red-dark': {
    borderColor: '#cc1f1a' },

  'border-red': {
    borderColor: '#e3342f' },

  'border-red-light': {
    borderColor: '#ef5753' },

  'border-red-lighter': {
    borderColor: '#f9acaa' },

  'border-red-lightest': {
    borderColor: '#fcebea' },

  'border-orange-darkest': {
    borderColor: '#462a16' },

  'border-orange-darker': {
    borderColor: '#613b1f' },

  'border-orange-dark': {
    borderColor: '#de751f' },

  'border-orange': {
    borderColor: '#f6993f' },

  'border-orange-light': {
    borderColor: '#faad63' },

  'border-orange-lighter': {
    borderColor: '#fcd9b6' },

  'border-orange-lightest': {
    borderColor: '#fff5eb' },

  'border-yellow-darkest': {
    borderColor: '#453411' },

  'border-yellow-darker': {
    borderColor: '#684f1d' },

  'border-yellow-dark': {
    borderColor: '#f2d024' },

  'border-yellow': {
    borderColor: '#ffed4a' },

  'border-yellow-light': {
    borderColor: '#fff382' },

  'border-yellow-lighter': {
    borderColor: '#fff9c2' },

  'border-yellow-lightest': {
    borderColor: '#fcfbeb' },

  'border-green-darkest': {
    borderColor: '#0f2f21' },

  'border-green-darker': {
    borderColor: '#1a4731' },

  'border-green-dark': {
    borderColor: '#1f9d55' },

  'border-green': {
    borderColor: '#38c172' },

  'border-green-light': {
    borderColor: '#51d88a' },

  'border-green-lighter': {
    borderColor: '#a2f5bf' },

  'border-green-lightest': {
    borderColor: '#e3fcec' },

  'border-teal-darkest': {
    borderColor: '#0d3331' },

  'border-teal-darker': {
    borderColor: '#20504f' },

  'border-teal-dark': {
    borderColor: '#38a89d' },

  'border-teal': {
    borderColor: '#4dc0b5' },

  'border-teal-light': {
    borderColor: '#64d5ca' },

  'border-teal-lighter': {
    borderColor: '#a0f0ed' },

  'border-teal-lightest': {
    borderColor: '#e8fffe' },

  'border-blue-darkest': {
    borderColor: '#12283a' },

  'border-blue-darker': {
    borderColor: '#1c3d5a' },

  'border-blue-dark': {
    borderColor: '#2779bd' },

  'border-blue': {
    borderColor: '#3490dc' },

  'border-blue-light': {
    borderColor: '#6cb2eb' },

  'border-blue-lighter': {
    borderColor: '#bcdefa' },

  'border-blue-lightest': {
    borderColor: '#eff8ff' },

  'border-indigo-darkest': {
    borderColor: '#191e38' },

  'border-indigo-darker': {
    borderColor: '#2f365f' },

  'border-indigo-dark': {
    borderColor: '#5661b3' },

  'border-indigo': {
    borderColor: '#6574cd' },

  'border-indigo-light': {
    borderColor: '#7886d7' },

  'border-indigo-lighter': {
    borderColor: '#b2b7ff' },

  'border-indigo-lightest': {
    borderColor: '#e6e8ff' },

  'border-purple-darkest': {
    borderColor: '#21183c' },

  'border-purple-darker': {
    borderColor: '#382b5f' },

  'border-purple-dark': {
    borderColor: '#794acf' },

  'border-purple': {
    borderColor: '#9561e2' },

  'border-purple-light': {
    borderColor: '#a779e9' },

  'border-purple-lighter': {
    borderColor: '#d6bbfc' },

  'border-purple-lightest': {
    borderColor: '#f3ebff' },

  'border-pink-darkest': {
    borderColor: '#451225' },

  'border-pink-darker': {
    borderColor: '#6f213f' },

  'border-pink-dark': {
    borderColor: '#eb5286' },

  'border-pink': {
    borderColor: '#f66d9b' },

  'border-pink-light': {
    borderColor: '#fa7ea8' },

  'border-pink-lighter': {
    borderColor: '#ffbbca' },

  'border-pink-lightest': {
    borderColor: '#ffebef' },

  'rounded-none': {
    borderRadius: 0 },

  'rounded-sm': {
    borderRadius: 2 },

  rounded: {
    borderRadius: 4 },

  'rounded-lg': {
    borderRadius: 8 },

  'rounded-full': {
    borderRadius: 9999 },

  'rounded-t-none': {
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0 },

  'rounded-r-none': {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0 },

  'rounded-b-none': {
    borderBottomRightRadius: 0,
    borderBottomLeftRadius: 0 },

  'rounded-l-none': {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0 },

  'rounded-t-sm': {
    borderTopLeftRadius: 2,
    borderTopRightRadius: 2 },

  'rounded-r-sm': {
    borderTopRightRadius: 2,
    borderBottomRightRadius: 2 },

  'rounded-b-sm': {
    borderBottomRightRadius: 2,
    borderBottomLeftRadius: 2 },

  'rounded-l-sm': {
    borderTopLeftRadius: 2,
    borderBottomLeftRadius: 2 },

  'rounded-t': {
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4 },

  'rounded-r': {
    borderTopRightRadius: 4,
    borderBottomRightRadius: 4 },

  'rounded-b': {
    borderBottomRightRadius: 4,
    borderBottomLeftRadius: 4 },

  'rounded-l': {
    borderTopLeftRadius: 4,
    borderBottomLeftRadius: 4 },

  'rounded-t-lg': {
    borderTopLeftRadius: 8,
    borderTopRightRadius: 8 },

  'rounded-r-lg': {
    borderTopRightRadius: 8,
    borderBottomRightRadius: 8 },

  'rounded-b-lg': {
    borderBottomRightRadius: 8,
    borderBottomLeftRadius: 8 },

  'rounded-l-lg': {
    borderTopLeftRadius: 8,
    borderBottomLeftRadius: 8 },

  'rounded-t-full': {
    borderTopLeftRadius: 9999,
    borderTopRightRadius: 9999 },

  'rounded-r-full': {
    borderTopRightRadius: 9999,
    borderBottomRightRadius: 9999 },

  'rounded-b-full': {
    borderBottomRightRadius: 9999,
    borderBottomLeftRadius: 9999 },

  'rounded-l-full': {
    borderTopLeftRadius: 9999,
    borderBottomLeftRadius: 9999 },

  'rounded-tl-none': {
    borderTopLeftRadius: 0 },

  'rounded-tr-none': {
    borderTopRightRadius: 0 },

  'rounded-br-none': {
    borderBottomRightRadius: 0 },

  'rounded-bl-none': {
    borderBottomLeftRadius: 0 },

  'rounded-tl-sm': {
    borderTopLeftRadius: 2 },

  'rounded-tr-sm': {
    borderTopRightRadius: 2 },

  'rounded-br-sm': {
    borderBottomRightRadius: 2 },

  'rounded-bl-sm': {
    borderBottomLeftRadius: 2 },

  'rounded-tl': {
    borderTopLeftRadius: 4 },

  'rounded-tr': {
    borderTopRightRadius: 4 },

  'rounded-br': {
    borderBottomRightRadius: 4 },

  'rounded-bl': {
    borderBottomLeftRadius: 4 },

  'rounded-tl-lg': {
    borderTopLeftRadius: 8 },

  'rounded-tr-lg': {
    borderTopRightRadius: 8 },

  'rounded-br-lg': {
    borderBottomRightRadius: 8 },

  'rounded-bl-lg': {
    borderBottomLeftRadius: 8 },

  'rounded-tl-full': {
    borderTopLeftRadius: 9999 },

  'rounded-tr-full': {
    borderTopRightRadius: 9999 },

  'rounded-br-full': {
    borderBottomRightRadius: 9999 },

  'rounded-bl-full': {
    borderBottomLeftRadius: 9999 },

  'border-solid': {
    borderStyle: 'solid' },

  'border-dashed': {
    borderStyle: 'dashed' },

  'border-dotted': {
    borderStyle: 'dotted' },

  'border-0': {
    borderWidth: 0 },

  'border-2': {
    borderWidth: 2 },

  'border-4': {
    borderWidth: 4 },

  'border-8': {
    borderWidth: 8 },

  border: {
    borderWidth: 1 },

  'border-t-0': {
    borderTopWidth: 0 },

  'border-r-0': {
    borderRightWidth: 0 },

  'border-b-0': {
    borderBottomWidth: 0 },

  'border-l-0': {
    borderLeftWidth: 0 },

  'border-t-2': {
    borderTopWidth: 2 },

  'border-r-2': {
    borderRightWidth: 2 },

  'border-b-2': {
    borderBottomWidth: 2 },

  'border-l-2': {
    borderLeftWidth: 2 },

  'border-t-4': {
    borderTopWidth: 4 },

  'border-r-4': {
    borderRightWidth: 4 },

  'border-b-4': {
    borderBottomWidth: 4 },

  'border-l-4': {
    borderLeftWidth: 4 },

  'border-t-8': {
    borderTopWidth: 8 },

  'border-r-8': {
    borderRightWidth: 8 },

  'border-b-8': {
    borderBottomWidth: 8 },

  'border-l-8': {
    borderLeftWidth: 8 },

  'border-t': {
    borderTopWidth: 1 },

  'border-r': {
    borderRightWidth: 1 },

  'border-b': {
    borderBottomWidth: 1 },

  'border-l': {
    borderLeftWidth: 1 },

  hidden: {
    display: 'none' },

  flex: {
    display: 'flex' },

  'flex-row': {
    flexDirection: 'row' },

  'flex-row-reverse': {
    flexDirection: 'row-reverse' },

  'flex-col': {
    flexDirection: 'column' },

  'flex-col-reverse': {
    flexDirection: 'column-reverse' },

  'flex-wrap': {
    flexWrap: 'wrap' },

  'flex-no-wrap': {
    flexWrap: 'nowrap' },

  'items-start': {
    alignItems: 'flex-start' },

  'items-end': {
    alignItems: 'flex-end' },

  'items-center': {
    alignItems: 'center' },

  'items-baseline': {
    alignItems: 'baseline' },

  'items-stretch': {
    alignItems: 'stretch' },

  'self-auto': {
    alignSelf: 'auto' },

  'self-start': {
    alignSelf: 'flex-start' },

  'self-end': {
    alignSelf: 'flex-end' },

  'self-center': {
    alignSelf: 'center' },

  'self-stretch': {
    alignSelf: 'stretch' },

  'justify-start': {
    justifyContent: 'flex-start' },

  'justify-end': {
    justifyContent: 'flex-end' },

  'justify-center': {
    justifyContent: 'center' },

  'justify-between': {
    justifyContent: 'space-between' },

  'justify-around': {
    justifyContent: 'space-around' },

  'content-center': {
    alignContent: 'center' },

  'content-start': {
    alignContent: 'flex-start' },

  'content-end': {
    alignContent: 'flex-end' },

  'content-between': {
    alignContent: 'space-between' },

  'content-around': {
    alignContent: 'space-around' },

  'flex-1': {
    flex: 1 },

  'font-sans': {
    fontFamily: 'system-ui, BlinkMacSystemFont, -apple-system, Segoe UI, Roboto, Oxygen, Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif' },

  'font-serif': {
    fontFamily: 'Constantia, Lucida Bright, Lucidabright, Lucida Serif, Lucida, DejaVu Serif, Bitstream Vera Serif, Liberation Serif, Georgia, serif' },

  'font-mono': {
    fontFamily: 'Menlo, Monaco, Consolas, Liberation Mono, Courier New, monospace' },

  'font-hairline': {
    fontWeight: '100' },

  'font-thin': {
    fontWeight: '200' },

  'font-light': {
    fontWeight: '300' },

  'font-normal': {
    fontWeight: '400' },

  'font-medium': {
    fontWeight: '500' },

  'font-semibold': {
    fontWeight: '600' },

  'font-bold': {
    fontWeight: '700' },

  'font-extrabold': {
    fontWeight: '800' },

  'font-black': {
    fontWeight: '900' },

  'h-1': {
    height: 4 },

  'h-2': {
    height: 8 },

  'h-3': {
    height: 12 },

  'h-4': {
    height: 16 },

  'h-6': {
    height: 24 },

  'h-8': {
    height: 32 },

  'h-10': {
    height: 40 },

  'h-12': {
    height: 48 },

  'h-16': {
    height: 64 },

  'h-24': {
    height: 96 },

  'h-32': {
    height: 128 },

  'h-48': {
    height: 192 },

  'h-64': {
    height: 256 },

  'h-auto': {
    height: 'auto' },

  'h-px': {
    height: 1 },

  'h-full': {
    height: '100%' },

  'leading-none': {
    lineHeight: 1 },

  'leading-tight': {
    lineHeight: 1 },

  'leading-normal': {
    lineHeight: 1 },

  'leading-loose': {
    lineHeight: 2 },

  'm-0': {
    marginTop: 0,
    marginBottom: 0,
    marginRight: 0,
    marginLeft: 0 },

  'm-1': {
    marginTop: 4,
    marginBottom: 4,
    marginRight: 4,
    marginLeft: 4 },

  'm-2': {
    marginTop: 8,
    marginBottom: 8,
    marginRight: 8,
    marginLeft: 8 },

  'm-3': {
    marginTop: 12,
    marginBottom: 12,
    marginRight: 12,
    marginLeft: 12 },

  'm-4': {
    marginTop: 16,
    marginBottom: 16,
    marginRight: 16,
    marginLeft: 16 },

  'm-6': {
    marginTop: 24,
    marginBottom: 24,
    marginRight: 24,
    marginLeft: 24 },

  'm-8': {
    marginTop: 32,
    marginBottom: 32,
    marginRight: 32,
    marginLeft: 32 },

  'm-auto': {
    marginTop: 'auto',
    marginBottom: 'auto',
    marginRight: 'auto',
    marginLeft: 'auto' },

  'm-px': {
    marginTop: 1,
    marginBottom: 1,
    marginRight: 1,
    marginLeft: 1 },

  'my-0': {
    marginTop: 0,
    marginBottom: 0 },

  'mx-0': {
    marginLeft: 0,
    marginRight: 0 },

  'my-1': {
    marginTop: 4,
    marginBottom: 4 },

  'mx-1': {
    marginLeft: 4,
    marginRight: 4 },

  'my-2': {
    marginTop: 8,
    marginBottom: 8 },

  'mx-2': {
    marginLeft: 8,
    marginRight: 8 },

  'my-3': {
    marginTop: 12,
    marginBottom: 12 },

  'mx-3': {
    marginLeft: 12,
    marginRight: 12 },

  'my-4': {
    marginTop: 16,
    marginBottom: 16 },

  'mx-4': {
    marginLeft: 16,
    marginRight: 16 },

  'my-6': {
    marginTop: 24,
    marginBottom: 24 },

  'mx-6': {
    marginLeft: 24,
    marginRight: 24 },

  'my-8': {
    marginTop: 32,
    marginBottom: 32 },

  'mx-8': {
    marginLeft: 32,
    marginRight: 32 },

  'my-auto': {
    marginTop: 'auto',
    marginBottom: 'auto' },

  'mx-auto': {
    marginLeft: 'auto',
    marginRight: 'auto' },

  'my-px': {
    marginTop: 1,
    marginBottom: 1 },

  'mx-px': {
    marginLeft: 1,
    marginRight: 1 },

  'mt-0': {
    marginTop: 0 },

  'mr-0': {
    marginRight: 0 },

  'mb-0': {
    marginBottom: 0 },

  'ml-0': {
    marginLeft: 0 },

  'mt-1': {
    marginTop: 4 },

  'mr-1': {
    marginRight: 4 },

  'mb-1': {
    marginBottom: 4 },

  'ml-1': {
    marginLeft: 4 },

  'mt-2': {
    marginTop: 8 },

  'mr-2': {
    marginRight: 8 },

  'mb-2': {
    marginBottom: 8 },

  'ml-2': {
    marginLeft: 8 },

  'mt-3': {
    marginTop: 12 },

  'mr-3': {
    marginRight: 12 },

  'mb-3': {
    marginBottom: 12 },

  'ml-3': {
    marginLeft: 12 },

  'mt-4': {
    marginTop: 16 },

  'mr-4': {
    marginRight: 16 },

  'mb-4': {
    marginBottom: 16 },

  'ml-4': {
    marginLeft: 16 },

  'mt-6': {
    marginTop: 24 },

  'mr-6': {
    marginRight: 24 },

  'mb-6': {
    marginBottom: 24 },

  'ml-6': {
    marginLeft: 24 },

  'mt-8': {
    marginTop: 32 },

  'mr-8': {
    marginRight: 32 },

  'mb-8': {
    marginBottom: 32 },

  'ml-8': {
    marginLeft: 32 },

  'mt-auto': {
    marginTop: 'auto' },

  'mr-auto': {
    marginRight: 'auto' },

  'mb-auto': {
    marginBottom: 'auto' },

  'ml-auto': {
    marginLeft: 'auto' },

  'mt-px': {
    marginTop: 1 },

  'mr-px': {
    marginRight: 1 },

  'mb-px': {
    marginBottom: 1 },

  'ml-px': {
    marginLeft: 1 },

  'max-h-full': {
    maxHeight: '100%' },

  'max-w-xs': {
    maxWidth: 320 },

  'max-w-sm': {
    maxWidth: 480 },

  'max-w-md': {
    maxWidth: 640 },

  'max-w-lg': {
    maxWidth: 800 },

  'max-w-xl': {
    maxWidth: 960 },

  'max-w-2xl': {
    maxWidth: 1120 },

  'max-w-3xl': {
    maxWidth: 1280 },

  'max-w-4xl': {
    maxWidth: 1440 },

  'max-w-5xl': {
    maxWidth: 1600 },

  'max-w-full': {
    maxWidth: '100%' },

  'min-h-0': {
    minHeight: 0 },

  'min-h-full': {
    minHeight: '100%' },

  'min-w-0': {
    minWidth: 0 },

  'min-w-full': {
    minWidth: '100%' },

  '-m-0': {
    marginTop: 0,
    marginBottom: 0,
    marginRight: 0,
    marginLeft: 0 },

  '-m-1': {
    marginTop: -4,
    marginBottom: -4,
    marginRight: -4,
    marginLeft: -4 },

  '-m-2': {
    marginTop: -8,
    marginBottom: -8,
    marginRight: -8,
    marginLeft: -8 },

  '-m-3': {
    marginTop: -12,
    marginBottom: -12,
    marginRight: -12,
    marginLeft: -12 },

  '-m-4': {
    marginTop: -16,
    marginBottom: -16,
    marginRight: -16,
    marginLeft: -16 },

  '-m-6': {
    marginTop: -24,
    marginBottom: -24,
    marginRight: -24,
    marginLeft: -24 },

  '-m-8': {
    marginTop: -32,
    marginBottom: -32,
    marginRight: -32,
    marginLeft: -32 },

  '-m-px': {
    marginTop: -1,
    marginBottom: -1,
    marginRight: -1,
    marginLeft: -1 },

  '-my-0': {
    marginTop: 0,
    marginBottom: 0 },

  '-mx-0': {
    marginLeft: 0,
    marginRight: 0 },

  '-my-1': {
    marginTop: -4,
    marginBottom: -4 },

  '-mx-1': {
    marginLeft: -4,
    marginRight: -4 },

  '-my-2': {
    marginTop: -8,
    marginBottom: -8 },

  '-mx-2': {
    marginLeft: -8,
    marginRight: -8 },

  '-my-3': {
    marginTop: -12,
    marginBottom: -12 },

  '-mx-3': {
    marginLeft: -12,
    marginRight: -12 },

  '-my-4': {
    marginTop: -16,
    marginBottom: -16 },

  '-mx-4': {
    marginLeft: -16,
    marginRight: -16 },

  '-my-6': {
    marginTop: -24,
    marginBottom: -24 },

  '-mx-6': {
    marginLeft: -24,
    marginRight: -24 },

  '-my-8': {
    marginTop: -32,
    marginBottom: -32 },

  '-mx-8': {
    marginLeft: -32,
    marginRight: -32 },

  '-my-px': {
    marginTop: -1,
    marginBottom: -1 },

  '-mx-px': {
    marginLeft: -1,
    marginRight: -1 },

  '-mt-0': {
    marginTop: 0 },

  '-mr-0': {
    marginRight: 0 },

  '-mb-0': {
    marginBottom: 0 },

  '-ml-0': {
    marginLeft: 0 },

  '-mt-1': {
    marginTop: -4 },

  '-mr-1': {
    marginRight: -4 },

  '-mb-1': {
    marginBottom: -4 },

  '-ml-1': {
    marginLeft: -4 },

  '-mt-2': {
    marginTop: -8 },

  '-mr-2': {
    marginRight: -8 },

  '-mb-2': {
    marginBottom: -8 },

  '-ml-2': {
    marginLeft: -8 },

  '-mt-3': {
    marginTop: -12 },

  '-mr-3': {
    marginRight: -12 },

  '-mb-3': {
    marginBottom: -12 },

  '-ml-3': {
    marginLeft: -12 },

  '-mt-4': {
    marginTop: -16 },

  '-mr-4': {
    marginRight: -16 },

  '-mb-4': {
    marginBottom: -16 },

  '-ml-4': {
    marginLeft: -16 },

  '-mt-6': {
    marginTop: -24 },

  '-mr-6': {
    marginRight: -24 },

  '-mb-6': {
    marginBottom: -24 },

  '-ml-6': {
    marginLeft: -24 },

  '-mt-8': {
    marginTop: -32 },

  '-mr-8': {
    marginRight: -32 },

  '-mb-8': {
    marginBottom: -32 },

  '-ml-8': {
    marginLeft: -32 },

  '-mt-px': {
    marginTop: -1 },

  '-mr-px': {
    marginRight: -1 },

  '-mb-px': {
    marginBottom: -1 },

  '-ml-px': {
    marginLeft: -1 },

  'opacity-0': {
    opacity: 0 },

  'opacity-25': {
    opacity: 0.25 },

  'opacity-50': {
    opacity: 0.5 },

  'opacity-75': {
    opacity: 0.75 },

  'opacity-100': {
    opacity: 1 },

  'p-0': {
    paddingTop: 0,
    paddingBottom: 0,
    paddingRight: 0,
    paddingLeft: 0 },

  'p-1': {
    paddingTop: 4,
    paddingBottom: 4,
    paddingRight: 4,
    paddingLeft: 4 },

  'p-2': {
    paddingTop: 8,
    paddingBottom: 8,
    paddingRight: 8,
    paddingLeft: 8 },

  'p-3': {
    paddingTop: 12,
    paddingBottom: 12,
    paddingRight: 12,
    paddingLeft: 12 },

  'p-4': {
    paddingTop: 16,
    paddingBottom: 16,
    paddingRight: 16,
    paddingLeft: 16 },

  'p-6': {
    paddingTop: 24,
    paddingBottom: 24,
    paddingRight: 24,
    paddingLeft: 24 },

  'p-8': {
    paddingTop: 32,
    paddingBottom: 32,
    paddingRight: 32,
    paddingLeft: 32 },

  'p-px': {
    paddingTop: 1,
    paddingBottom: 1,
    paddingRight: 1,
    paddingLeft: 1 },

  'py-0': {
    paddingTop: 0,
    paddingBottom: 0 },

  'px-0': {
    paddingLeft: 0,
    paddingRight: 0 },

  'py-1': {
    paddingTop: 4,
    paddingBottom: 4 },

  'px-1': {
    paddingLeft: 4,
    paddingRight: 4 },

  'py-2': {
    paddingTop: 8,
    paddingBottom: 8 },

  'px-2': {
    paddingLeft: 8,
    paddingRight: 8 },

  'py-3': {
    paddingTop: 12,
    paddingBottom: 12 },

  'px-3': {
    paddingLeft: 12,
    paddingRight: 12 },

  'py-4': {
    paddingTop: 16,
    paddingBottom: 16 },

  'px-4': {
    paddingLeft: 16,
    paddingRight: 16 },

  'py-6': {
    paddingTop: 24,
    paddingBottom: 24 },

  'px-6': {
    paddingLeft: 24,
    paddingRight: 24 },

  'py-8': {
    paddingTop: 32,
    paddingBottom: 32 },

  'px-8': {
    paddingLeft: 32,
    paddingRight: 32 },

  'py-px': {
    paddingTop: 1,
    paddingBottom: 1 },

  'px-px': {
    paddingLeft: 1,
    paddingRight: 1 },

  'pt-0': {
    paddingTop: 0 },

  'pr-0': {
    paddingRight: 0 },

  'pb-0': {
    paddingBottom: 0 },

  'pl-0': {
    paddingLeft: 0 },

  'pt-1': {
    paddingTop: 4 },

  'pr-1': {
    paddingRight: 4 },

  'pb-1': {
    paddingBottom: 4 },

  'pl-1': {
    paddingLeft: 4 },

  'pt-2': {
    paddingTop: 8 },

  'pr-2': {
    paddingRight: 8 },

  'pb-2': {
    paddingBottom: 8 },

  'pl-2': {
    paddingLeft: 8 },

  'pt-3': {
    paddingTop: 12 },

  'pr-3': {
    paddingRight: 12 },

  'pb-3': {
    paddingBottom: 12 },

  'pl-3': {
    paddingLeft: 12 },

  'pt-4': {
    paddingTop: 16 },

  'pr-4': {
    paddingRight: 16 },

  'pb-4': {
    paddingBottom: 16 },

  'pl-4': {
    paddingLeft: 16 },

  'pt-6': {
    paddingTop: 24 },

  'pr-6': {
    paddingRight: 24 },

  'pb-6': {
    paddingBottom: 24 },

  'pl-6': {
    paddingLeft: 24 },

  'pt-8': {
    paddingTop: 32 },

  'pr-8': {
    paddingRight: 32 },

  'pb-8': {
    paddingBottom: 32 },

  'pl-8': {
    paddingLeft: 32 },

  'pt-px': {
    paddingTop: 1 },

  'pr-px': {
    paddingRight: 1 },

  'pb-px': {
    paddingBottom: 1 },

  'pl-px': {
    paddingLeft: 1 },

  absolute: {
    position: 'absolute' },

  relative: {
    position: 'relative' },

  'pin-none': {
    top: 'auto',
    right: 'auto',
    bottom: 'auto',
    left: 'auto' },

  pin: {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0 },

  'pin-y': {
    top: 0,
    bottom: 0 },

  'pin-x': {
    right: 0,
    left: 0 },

  'pin-t': {
    top: 0 },

  'pin-r': {
    right: 0 },

  'pin-b': {
    bottom: 0 },

  'pin-l': {
    left: 0 },

  'text-left': {
    textAlign: 'left' },

  'text-center': {
    textAlign: 'center' },

  'text-right': {
    textAlign: 'right' },

  'text-justify': {
    textAlign: 'justify' },

  'text-transparent': {
    color: 'transparent' },

  'text-black': {
    color: '#22292f' },

  'text-grey-darkest': {
    color: '#3d4852' },

  'text-grey-darker': {
    color: '#606f7b' },

  'text-grey-dark': {
    color: '#8795a1' },

  'text-grey': {
    color: '#b8c2cc' },

  'text-grey-light': {
    color: '#dae1e7' },

  'text-grey-lighter': {
    color: '#f1f5f8' },

  'text-grey-lightest': {
    color: '#f8fafc' },

  'text-white': {
    color: '#fff' },

  'text-red-darkest': {
    color: '#3b0d0c' },

  'text-red-darker': {
    color: '#621b18' },

  'text-red-dark': {
    color: '#cc1f1a' },

  'text-red': {
    color: '#e3342f' },

  'text-red-light': {
    color: '#ef5753' },

  'text-red-lighter': {
    color: '#f9acaa' },

  'text-red-lightest': {
    color: '#fcebea' },

  'text-orange-darkest': {
    color: '#462a16' },

  'text-orange-darker': {
    color: '#613b1f' },

  'text-orange-dark': {
    color: '#de751f' },

  'text-orange': {
    color: '#f6993f' },

  'text-orange-light': {
    color: '#faad63' },

  'text-orange-lighter': {
    color: '#fcd9b6' },

  'text-orange-lightest': {
    color: '#fff5eb' },

  'text-yellow-darkest': {
    color: '#453411' },

  'text-yellow-darker': {
    color: '#684f1d' },

  'text-yellow-dark': {
    color: '#f2d024' },

  'text-yellow': {
    color: '#ffed4a' },

  'text-yellow-light': {
    color: '#fff382' },

  'text-yellow-lighter': {
    color: '#fff9c2' },

  'text-yellow-lightest': {
    color: '#fcfbeb' },

  'text-green-darkest': {
    color: '#0f2f21' },

  'text-green-darker': {
    color: '#1a4731' },

  'text-green-dark': {
    color: '#1f9d55' },

  'text-green': {
    color: '#38c172' },

  'text-green-light': {
    color: '#51d88a' },

  'text-green-lighter': {
    color: '#a2f5bf' },

  'text-green-lightest': {
    color: '#e3fcec' },

  'text-teal-darkest': {
    color: '#0d3331' },

  'text-teal-darker': {
    color: '#20504f' },

  'text-teal-dark': {
    color: '#38a89d' },

  'text-teal': {
    color: '#4dc0b5' },

  'text-teal-light': {
    color: '#64d5ca' },

  'text-teal-lighter': {
    color: '#a0f0ed' },

  'text-teal-lightest': {
    color: '#e8fffe' },

  'text-blue-darkest': {
    color: '#12283a' },

  'text-blue-darker': {
    color: '#1c3d5a' },

  'text-blue-dark': {
    color: '#2779bd' },

  'text-blue': {
    color: '#3490dc' },

  'text-blue-light': {
    color: '#6cb2eb' },

  'text-blue-lighter': {
    color: '#bcdefa' },

  'text-blue-lightest': {
    color: '#eff8ff' },

  'text-indigo-darkest': {
    color: '#191e38' },

  'text-indigo-darker': {
    color: '#2f365f' },

  'text-indigo-dark': {
    color: '#5661b3' },

  'text-indigo': {
    color: '#6574cd' },

  'text-indigo-light': {
    color: '#7886d7' },

  'text-indigo-lighter': {
    color: '#b2b7ff' },

  'text-indigo-lightest': {
    color: '#e6e8ff' },

  'text-purple-darkest': {
    color: '#21183c' },

  'text-purple-darker': {
    color: '#382b5f' },

  'text-purple-dark': {
    color: '#794acf' },

  'text-purple': {
    color: '#9561e2' },

  'text-purple-light': {
    color: '#a779e9' },

  'text-purple-lighter': {
    color: '#d6bbfc' },

  'text-purple-lightest': {
    color: '#f3ebff' },

  'text-pink-darkest': {
    color: '#451225' },

  'text-pink-darker': {
    color: '#6f213f' },

  'text-pink-dark': {
    color: '#eb5286' },

  'text-pink': {
    color: '#f66d9b' },

  'text-pink-light': {
    color: '#fa7ea8' },

  'text-pink-lighter': {
    color: '#ffbbca' },

  'text-pink-lightest': {
    color: '#ffebef' },

  'text-xs': {
    fontSize: 12 },

  'text-sm': {
    fontSize: 14 },

  'text-base': {
    fontSize: 16 },

  'text-lg': {
    fontSize: 18 },

  'text-xl': {
    fontSize: 20 },

  'text-2xl': {
    fontSize: 24 },

  'text-3xl': {
    fontSize: 30 },

  'text-4xl': {
    fontSize: 36 },

  'text-5xl': {
    fontSize: 48 },

  italic: {
    fontStyle: 'italic' },

  roman: {
    fontStyle: 'normal' },

  'tracking-tight': {
    letterSpacing: -0.8 },

  'tracking-normal': {
    letterSpacing: 0 },

  'tracking-wide': {
    letterSpacing: 0.8 },

  'w-1': {
    width: 4 },

  'w-2': {
    width: 8 },

  'w-3': {
    width: 12 },

  'w-4': {
    width: 16 },

  'w-6': {
    width: 24 },

  'w-8': {
    width: 32 },

  'w-10': {
    width: 40 },

  'w-12': {
    width: 48 },

  'w-16': {
    width: 64 },

  'w-24': {
    width: 96 },

  'w-32': {
    width: 128 },

  'w-48': {
    width: 192 },

  'w-64': {
    width: 256 },

  'w-auto': {
    width: 'auto' },

  'w-px': {
    width: 1 },

  'w-1/2': {
    width: '50%' },

  'w-1/3': {
    width: '33.33333%' },

  'w-2/3': {
    width: '66.66667%' },

  'w-1/4': {
    width: '25%' },

  'w-3/4': {
    width: '75%' },

  'w-1/5': {
    width: '20%' },

  'w-2/5': {
    width: '40%' },

  'w-3/5': {
    width: '60%' },

  'w-4/5': {
    width: '80%' },

  'w-1/6': {
    width: '16.66667%' },

  'w-5/6': {
    width: '83.33333%' },

  'w-full': {
    width: '100%' },

  'z-1': {
    zIndex: 4 },

  'z-2': {
    zIndex: 8 },

  'z-3': {
    zIndex: 12 },

  'z-4': {
    zIndex: 16 },

  'z-5': {
    zIndex: 20 } });

var getTailwindStyle = function (tailwind) {
  return typeof tailwind === 'string' ? tailwindStyles[tailwind] : tailwind.map(function (style) {return tailwindStyles[style];});
};

var getTachyonsStyle = function (tachyons) {
  return typeof tachyons === 'string' ? tachyonsStyles[tachyons] : tachyons.map(function (style) {return tachyonsStyles[style];});
};

function getDisplayNameHOC(WrappedComponent) {
  var _a;
  return ((_a = WrappedComponent.displayName) !== null && _a !== void 0 ? _a : WrappedComponent.name) || 'Component';
}

var isIOS = Platform.OS === 'ios';
var PlatformIOS = Platform;
var isIpad = PlatformIOS.isPad;
var isSmallDevice = SCREEN_HEIGHT < 750;

var fontWeightTitle = isIOS ? { fontWeight: '500' } : { fontFamily: 'sans-serif-medium', fontWeight: 'normal' };
var fontWeightText = isIOS ? { fontWeight: '400' } : { fontFamily: 'sans-serif', fontWeight: 'normal' };

var CONSTANT_LINE_HEIGHT = 1.37;
var CONSTANT_H1 = 2.2;
var CONSTANT_H2 = 2;
var CONSTANT_H3 = 1.75;
var CONSTANT_H4 = 1.6;
var CONSTANT_H5 = 1.4;
var CONSTANT_H6 = 1.3;
var CONSTANT_H7 = 1;
var CONSTANT_H8 = 0.8;
var CONSTANT_SMALL = 0.8;
function _getTextStyle(type, sizes, colors, colorNative, color) {
  var font = sizes.font;
  var headingColor = !!colorNative ? colorNative : !!color ? colors[color] : colors.dark1;
  var textColor = !!colorNative ? colorNative : !!color ? colors[color] : colors.dark2;
  switch (type) {
    case 'h1':
      return __assign(__assign({ fontSize: font * CONSTANT_H1 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H1, color: headingColor });
    case 'h2':
      return __assign(__assign({ fontSize: font * CONSTANT_H2 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H2, color: headingColor });
    case 'h3':
      return __assign(__assign({ fontSize: font * CONSTANT_H3 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H3, color: headingColor });
    case 'h4':
      return __assign(__assign({ fontSize: font * CONSTANT_H4 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H4, color: headingColor });
    case 'h5':
      return __assign(__assign({ fontSize: font * CONSTANT_H5 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H5, color: headingColor });
    case 'h6':
      return __assign(__assign({ fontSize: font * CONSTANT_H6 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H6, color: headingColor });
    case 'h7':
      return __assign(__assign({ fontSize: font * CONSTANT_H7 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H7, color: headingColor });
    case 'h8':
      return __assign(__assign({ fontSize: font * CONSTANT_H8 }, fontWeightTitle), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_H8, color: headingColor });
    case 'p':
      return __assign(__assign({ fontSize: font }, fontWeightText), { lineHeight: font * CONSTANT_LINE_HEIGHT, color: textColor });
    case 'i':
      return __assign(__assign({ fontSize: font }, fontWeightText), { lineHeight: font * CONSTANT_LINE_HEIGHT, color: textColor, fontStyle: 'italic' });
    case 'small':
      return __assign(__assign({ fontSize: font * CONSTANT_SMALL }, fontWeightText), { lineHeight: font * CONSTANT_LINE_HEIGHT * CONSTANT_SMALL, color: textColor });
    default:
      return {};}

}
function withTextStyles(Component, styleProp) {
  if (styleProp === void 0) {styleProp = 'style';}
  var WithTextStyles = /*#__PURE__*/forwardRef(function (_a, ref) {
    var _b;
    var _c = _a.tailwind,tailwind = _c === void 0 ? [] : _c,_d = _a.tachyons,tachyons = _d === void 0 ? [] : _d,_e = _a.type,type = _e === void 0 ? 'p' : _e,size = _a.size,color = _a.color,_f = _a.colorNative,colorNative = _f === void 0 ? '' : _f,_g = _a.backgroundColor,backgroundColor = _g === void 0 ? 'transparent' : _g,_h = _a.style,style = _h === void 0 ? {} : _h,rest = __rest(_a, ["tailwind", "tachyons", "type", "size", "color", "colorNative", "backgroundColor", "style"]);
    var _j = useTheme(),sizes = _j.sizes,colors = _j.colors;
    var inlineStyle = [
    { backgroundColor: !!colors[backgroundColor] ? colors[backgroundColor] : backgroundColor },
    !!size ? { fontSize: size } : {}];

    var textStyle = _getTextStyle(type, sizes, colors, colorNative, color);
    var styles = [textStyle, inlineStyle, getTailwindStyle(tailwind), getTachyonsStyle(tachyons), style];
    return /*#__PURE__*/React.createElement(Component, __assign({ ref: ref }, rest, (_b = {}, _b[styleProp] = styles, _b)));
  });
  WithTextStyles.displayName = "WithTextStyles(" + getDisplayNameHOC(Component) + ")";
  return WithTextStyles;
}

var styles$h = StyleSheet.create({
  debug: {
    borderWidth: 1,
    borderColor: '#eb2226' } });


var TextWithStyles = withTextStyles(Text$1);
var TextComponent = /*#__PURE__*/forwardRef(function (_a, ref) {
  var children = _a.children,style = _a.style,rest = __rest(_a, ["children", "style"]);
  var debug = useTheme().debug;
  return /*#__PURE__*/React.createElement(TextWithStyles, __assign({}, rest, { style: [style, debug ? styles$h.debug : {}], ref: ref }), children);
});
var Text = /*#__PURE__*/memo(TextComponent);

function getFlexStyle(flex) {
  if (typeof flex === 'boolean') {
    return flex ? tachyonsStyles.flex : {};
  }
  if (typeof flex === 'number') {
    return { flex: flex };
  }
  return {};
}

function getFlexWrapStyle(flexWrap) {
  switch (flexWrap) {
    case 'nowrap':
      return tachyonsStyles.flexNowrap;
    case 'wrap-reverse':
      return tachyonsStyles.flexWrapReverse;
    case 'wrap':
      return tachyonsStyles.flexWrap;
    default:
      return tachyonsStyles.empty;}

}

function getJustifyContentStyle(justifyContent) {
  switch (justifyContent) {
    case 'flex-start':
      return tachyonsStyles.justifyStart;
    case 'flex-end':
      return tachyonsStyles.justifyEnd;
    case 'center':
      return tachyonsStyles.justifyCenter;
    case 'space-around':
      return tachyonsStyles.justifyAround;
    case 'space-between':
      return tachyonsStyles.justifyBetween;
    case 'space-evenly':
      return tachyonsStyles.justifyEvenly;
    default:
      return tachyonsStyles.empty;}

}

var getColPercent = function (col) {return 100 * col / 6 + "%";};
var styles$g = StyleSheet.create({
  col6: {
    width: getColPercent(6) },

  col5: {
    width: getColPercent(5) },

  col4: {
    width: getColPercent(4) },

  col3: {
    width: getColPercent(3) },

  col2: {
    width: getColPercent(2) },

  col1: {
    width: getColPercent(1) },

  empty: {} });

function getColumnStyle(column) {
  switch (column) {
    case '1/6':
      return styles$g.col1;
    case '2/6':
      return styles$g.col2;
    case '3/6':
      return styles$g.col3;
    case '4/6':
      return styles$g.col4;
    case '5/6':
      return styles$g.col5;
    case '6/6':
      return styles$g.col6;
    default:
      return styles$g.empty;}

}

function getFlexDirectionStyle(flexDirection) {
  switch (flexDirection) {
    case 'column':
      return tachyonsStyles.flexColumn;
    case 'column-reverse':
      return tachyonsStyles.flexColumnReverse;
    case 'row':
      return [tachyonsStyles.flexRow, tachyonsStyles.flexWrap];
    case 'row-reverse':
      return tachyonsStyles.flexRowReverse;
    default:
      return tachyonsStyles.empty;}

}

function getAlignItemsStyle(alignItems) {
  switch (alignItems) {
    case 'flex-start':
      return tachyonsStyles.itemsStart;
    case 'flex-end':
      return tachyonsStyles.itemsEnd;
    case 'center':
      return tachyonsStyles.itemsCenter;
    case 'baseline':
      return tachyonsStyles.itemsBaseline;
    case 'stretch':
      return tachyonsStyles.itemsStretch;
    default:
      return {};}

}

function withViewStyles(Component, styleProp) {
  if (styleProp === void 0) {styleProp = 'style';}
  var WithViewStyles = /*#__PURE__*/forwardRef(function (_a, ref) {
    var _b;
    var _c = _a.tailwind,tailwind = _c === void 0 ? [] : _c,_d = _a.tachyons,tachyons = _d === void 0 ? [] : _d,_e = _a.backgroundColor,backgroundColor = _e === void 0 ? 'transparent' : _e,_f = _a.flex,flex = _f === void 0 ? false : _f,justifyContent = _a.justifyContent,alignItems = _a.alignItems,flexDirection = _a.flexDirection,column = _a.column,flexWrap = _a.flexWrap,_g = _a.style,style = _g === void 0 ? {} : _g,rest = __rest(_a, ["tailwind", "tachyons", "backgroundColor", "flex", "justifyContent", "alignItems", "flexDirection", "column", "flexWrap", "style"]);
    var colors = useTheme().colors;
    var flexStyle = getFlexStyle(flex);
    var justifyContentStyle = getJustifyContentStyle(justifyContent);
    var alignItemsStyle = getAlignItemsStyle(alignItems);
    var flexDirectionStyle = getFlexDirectionStyle(flexDirection);
    var columnStyle = !!column ? getColumnStyle(column) : {};
    var flexWrapStyle = getFlexWrapStyle(flexWrap);
    var inlineStyle = { backgroundColor: !!colors[backgroundColor] ? colors[backgroundColor] : backgroundColor };
    var styles = [
    flexStyle,
    justifyContentStyle,
    alignItemsStyle,
    flexDirectionStyle,
    flexWrapStyle,
    columnStyle,
    inlineStyle,
    getTailwindStyle(tailwind),
    getTachyonsStyle(tachyons),
    style];

    return /*#__PURE__*/React.createElement(Component, __assign({ ref: ref }, rest, (_b = {}, _b[styleProp] = styles, _b)));
  });
  WithViewStyles.displayName = "WithViewStyles(" + getDisplayNameHOC(Component) + ")";
  return WithViewStyles;
}

var styles$f = StyleSheet.create({
  empty: {},
  shadow: {
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 8 },
    shadowOpacity: 0.08,
    shadowRadius: 5,
    elevation: 3 },

  debug: {
    borderWidth: 1,
    borderColor: '#01295c' } });

var ViewWithStyles = withViewStyles(View$1);
var ViewComponent = /*#__PURE__*/forwardRef(function (_a, ref) {
  var children = _a.children,_b = _a.flex,flex = _b === void 0 ? false : _b,_c = _a.safeAreaView,safeAreaView = _c === void 0 ? false : _c,_d = _a.safeAreaViewBottom,safeAreaViewBottom = _d === void 0 ? false : _d,_e = _a.shadow,shadow = _e === void 0 ? false : _e,_f = _a.style,style = _f === void 0 ? {} : _f,onMount = _a.onMount,rest = __rest(_a, ["children", "flex", "safeAreaView", "safeAreaViewBottom", "shadow", "style", "onMount"]);
  var debug = useTheme().debug;
  var insets = useSafeArea();
  var flexStyle = getFlexStyle(flex);
  var shadowStyle = shadow ? styles$f.shadow : {};
  useMount(function () {
    onMount === null || onMount === void 0 ? void 0 : onMount();
  });
  return /*#__PURE__*/React.createElement(ViewWithStyles, __assign({ ref: ref, style: [shadowStyle, flexStyle, style, debug ? styles$f.debug : styles$f.empty] }, rest),
  safeAreaView && /*#__PURE__*/React.createElement(View$1, { style: { height: insets.top } }),
  children,
  safeAreaViewBottom && /*#__PURE__*/React.createElement(View$1, { style: { height: insets.bottom } }));
});
var View = /*#__PURE__*/memo(ViewComponent);

var styles$e = StyleSheet.create({
  // container: {
  //   ...StyleSheet.absoluteFillObject,
  //   zIndex: 1000,
  //   borderWidth: 3,
  // },
  // inner: {
  //   position: 'relative',
  //   marginVertical: 10,
  // },
  container: {
    position: 'absolute',
    left: 10,
    right: 10,
    zIndex: 1000 },

  toastInner: {
    padding: 5,
    borderRadius: 5 } });

var ToastEvent = new Emitter();
var defaultPayload = {
  duration: 2000,
  content: '',
  style: {},
  placement: 'bottom',
  animationType: 'fade' };

var defaultState = {
  visible: false,
  content: '',
  style: {},
  placement: 'bottom',
  animationType: 'fade',
  tailwind: [],
  tachyons: [] };

var getPlacementStyle = function (placement, height) {
  switch (placement) {
    case 'top':
      return {
        top: 10 };

    case 'center':
      return {
        top: '50%',
        marginTop: !!height ? -height / 2 : 0 };

    case 'bottom':
      return {
        bottom: 10 };

    default:
      return {};}

};
var Toast = {
  show: function (payload) {
    var _payload = __assign(__assign({}, defaultPayload), typeof payload === 'string' ? { content: payload } : payload);
    ToastEvent.emit('TOAST_SHOW', _payload);
  } };

var ToastUI = /*#__PURE__*/memo(function () {
  var _a = useState(defaultState.visible),visible = _a[0],setVisible = _a[1];
  var _b = useState(defaultState.content),content = _b[0],setContent = _b[1];
  var _c = useState(defaultState.style),style = _c[0],setStyle = _c[1];
  var _d = useState(defaultState.placement),placement = _d[0],setPlacement = _d[1];
  var _e = useState(defaultState.animationType),animationType = _e[0],setAnimationType = _e[1];
  var _f = useState(defaultState.tailwind),tailwind = _f[0],setTailwind = _f[1];
  var _g = useState(defaultState.tachyons),tachyons = _g[0],setTachyons = _g[1];
  var _h = useState(0),height = _h[0],setHeight = _h[1];
  var show = useRef(0);
  var timeout = useRef(0);
  var _j = usePortalAnimation({
    visible: visible,
    animationType: animationType,
    animationDuration: 200,
    animmationEasing: Easing.bezier(0.46, 0.79, 0.54, 0.98),
    elementHeight: height }),
  visibleState = _j.visibleState,animationStyle = _j.animationStyle;
  var handleLayoutContainer = function (event) {
    setHeight(event.nativeEvent.layout.height);
  };
  var handleShow = function (payload) {
    var _a;
    setVisible(true);
    setContent(payload.content);
    setStyle(payload.style);
    !!payload.placement && setPlacement(payload.placement);
    !!payload.animationType && setAnimationType(payload.animationType);
    !!payload.tailwind && setTailwind(payload.tailwind);
    !!payload.tachyons && setTachyons(payload.tachyons);
    (_a = payload.onOpenEnd) === null || _a === void 0 ? void 0 : _a.call(payload);
    timeout.current = setTimeout(function () {
      var _a;
      setVisible(false);
      (_a = payload.onCloseEnd) === null || _a === void 0 ? void 0 : _a.call(payload);
      clearTimeout(timeout.current);
    }, payload.duration);
  };
  useMount(function () {
    if (!visible) {
      show.current = ToastEvent.on('TOAST_SHOW', handleShow);
    }
  });
  useUnmount(function () {
    ToastEvent.off(show.current);
    clearTimeout(timeout.current);
  });
  if (!visibleState) {
    return null;
  }
  return /*#__PURE__*/React.createElement(View, { style: [styles$e.container, getPlacementStyle(placement, height)], onLayout: handleLayoutContainer }, /*#__PURE__*/
  React.createElement(View, { safeAreaView: true }, /*#__PURE__*/
  React.createElement(Animated.View, { style: [tachyonsStyles.flex, animationStyle] }, /*#__PURE__*/
  React.createElement(View, { style: [styles$e.toastInner, style], tailwind: tailwind, tachyons: tachyons, backgroundColor: "dark1" }, typeof content === 'string' ? /*#__PURE__*/React.createElement(Text, { color: "gray1" }, content) : content))));
});

var isAndroid = Platform.OS === 'android';

var styles$d = StyleSheet.create({
  container: __assign(__assign({}, StyleSheet.absoluteFillObject), { zIndex: 900 }) });

var Event = new Emitter();
var defaultProps = {
  visible: true };

var ModalBase = function (_a) {
  var onRequestClose = _a.onRequestClose,otherProps = __rest(_a, ["onRequestClose"]);
  var backPress = useRef(0);
  Event.emit('MODAL_MOUNT', otherProps);
  var handleBackPress = function () {
    onRequestClose === null || onRequestClose === void 0 ? void 0 : onRequestClose();
  };
  useMount(function () {
    backPress.current = Event.on('MODAL_BACK_PRESS', handleBackPress);
  });
  useUnmount(function () {
    Event.off(backPress.current);
  });
  return null;
};
ModalBase.defaultProps = defaultProps;
var ModalBaseUI = /*#__PURE__*/memo(function () {
  var mounted = useRef(0);
  var _a = useState(false),visible = _a[0],setVisible = _a[1];
  var _b = useState(null),children = _b[0],setChildren = _b[1];
  var _c = useState({}),style = _c[0],setStyle = _c[1];
  var handleBackPress = function () {
    Event.emit('MODAL_BACK_PRESS');
    return null;
  };
  var handleMount = function (props) {
    !!props.visible && setVisible(props.visible);
    setChildren(props.children);
    !!props.style && setStyle(props.style);
  };
  useMount(function () {
    if (isAndroid) {
      BackHandler.addEventListener('hardwareBackPress', handleBackPress);
    }
    mounted.current = Event.on('MODAL_MOUNT', handleMount);
  });
  useUnmount(function () {
    Event.off(mounted.current);
  });
  if (!visible) {
    return null;
  }
  return /*#__PURE__*/React.createElement(View, { style: styles$d.container }, /*#__PURE__*/
  React.createElement(View, { flex: true, style: style }, children));
});

var defaultTheme = {
  colors: defaultColors,
  sizes: defaultSizes,
  styled: getDefaultStyles(defaultColors),
  debug: false };

var ThemeContext = /*#__PURE__*/createContext(defaultTheme);
function ThemeProvider(_a) {
  var _b = _a.themeOverrides,themeOverrides = _b === void 0 ? defaultTheme : _b,children = _a.children;
  var colors = __assign(__assign({}, defaultColors), themeOverrides.colors);
  var sizes = __assign(__assign({}, defaultSizes), themeOverrides.sizes);
  var styled = __assign(__assign({}, getDefaultStyles(colors)), themeOverrides.styled);
  var debug = themeOverrides.debug || false;
  return /*#__PURE__*/React.createElement(SafeAreaProvider, null, /*#__PURE__*/
  React.createElement(ThemeContext.Provider, { value: __assign(__assign({}, themeOverrides), { colors: colors,
      sizes: sizes,
      styled: styled,
      debug: debug }) },
  children, /*#__PURE__*/
  React.createElement(ToastUI, null), /*#__PURE__*/
  React.createElement(ModalBaseUI, null)));
}
var ThemeConsumer = ThemeContext.Consumer;
function useTheme() {
  var theme = useContext(ThemeContext);
  return theme;
}
function withTheme(Component) {
  var WithTheme = function (_a) {
    var rest = __rest(_a, []);
    var theme = useTheme();
    return /*#__PURE__*/React.createElement(Component, __assign({}, rest, { theme: theme }));
  };
  WithTheme.displayName = "WithTheme " + getDisplayNameHOC(Component);
  return WithTheme;
}

function withTachyons(Component, styleProp) {
  if (styleProp === void 0) {styleProp = 'style';}
  var WithTachyons = /*#__PURE__*/forwardRef(function (_a, ref) {
    var _b;
    var _c = _a.tachyons,tachyons = _c === void 0 ? [] : _c,style = _a.style,rest = __rest(_a, ["tachyons", "style"]);
    return /*#__PURE__*/React.createElement(Component, __assign({ ref: ref }, rest, (_b = {}, _b[styleProp] = [getTachyonsStyle(tachyons), style], _b)));
  });
  WithTachyons.displayName = "WithTachyons " + getDisplayNameHOC(Component);
  return WithTachyons;
}

function withTailwind(Component, styleProp) {
  if (styleProp === void 0) {styleProp = 'style';}
  var WithTailwind = /*#__PURE__*/forwardRef(function (_a, ref) {
    var _b;
    var _c = _a.tailwind,tailwind = _c === void 0 ? [] : _c,style = _a.style,rest = __rest(_a, ["tailwind", "style"]);
    return /*#__PURE__*/React.createElement(Component, __assign({ ref: ref }, rest, (_b = {}, _b[styleProp] = [getTailwindStyle(tailwind), style], _b)));
  });
  WithTailwind.displayName = "WithTailwind " + getDisplayNameHOC(Component);
  return WithTailwind;
}

var ContainerComponent = function (_a) {
  var children = _a.children,safeAreaView = _a.safeAreaView,_b = _a.tachyons,tachyons = _b === void 0 ? [] : _b,flex = _a.flex,style = _a.style,maxWidth = _a.maxWidth,_c = _a.innerTachyons,innerTachyons = _c === void 0 ? [] : _c;
  var sizes = useTheme().sizes;
  return /*#__PURE__*/React.createElement(View, { flex: flex, safeAreaView: safeAreaView, tachyons: __spreadArray(['itemsCenter'], typeof tachyons === 'string' ? [tachyons] : tachyons), style: style }, /*#__PURE__*/
  React.createElement(View, { flex: flex, style: { maxWidth: maxWidth !== null && maxWidth !== void 0 ? maxWidth : sizes.container }, tachyons: __spreadArray(['w100'], typeof innerTachyons === 'string' ? [innerTachyons] : innerTachyons) }, children));
};
var Container = /*#__PURE__*/memo(ContainerComponent);

var styles$c = StyleSheet.create({
  container: {
    marginBottom: 15 },

  error: {
    marginTop: 4 } });

function FormItemComponnent(_a) {
  var _b = _a.label,label = _b === void 0 ? '' : _b,_c = _a.errorMessage,errorMessage = _c === void 0 ? '' : _c,_d = _a.required,required = _d === void 0 ? false : _d,children = _a.children;
  var styled = useTheme().styled;
  return /*#__PURE__*/React.createElement(View, { style: styles$c.container },
  !!label && /*#__PURE__*/React.createElement(Text, { tachyons: ['f6', 'mt1'], style: styled.colorDark3 },
  label,
  " ",
  required && /*#__PURE__*/React.createElement(Text, { style: styled.colorSecondary }, "*")),
  children,
  !!errorMessage && /*#__PURE__*/React.createElement(Text, { tachyons: "f6", style: [styled.colorDanger, styles$c.error] }, errorMessage));
}
var FormItem = /*#__PURE__*/memo(FormItemComponnent);

var styles$b = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 54,
    paddingHorizontal: 5 },

  item: {
    flexGrow: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    overflow: 'hidden' },

  left: {
    justifyContent: 'flex-start' },

  center: {
    flexGrow: 1.5 },

  right: {
    justifyContent: 'flex-end' },

  hasCenter: {
    flexBasis: 1 },

  debug: {
    borderWidth: 2,
    borderColor: 'red' } });

function HeaderBaseComponent(_a) {
  var Left = _a.Left,Center = _a.Center,Right = _a.Right,_b = _a.containerStyle,containerStyle = _b === void 0 ? {} : _b,_c = _a.backgroundColor,backgroundColor = _c === void 0 ? '' : _c;
  var colors = useTheme().colors;
  var itemHasCenterStyle = Center ? styles$b.hasCenter : {};
  return /*#__PURE__*/React.createElement(View, { style: [{ backgroundColor: backgroundColor || colors.light }, containerStyle], safeAreaView: true }, /*#__PURE__*/
  React.createElement(View, { style: styles$b.container }, /*#__PURE__*/
  React.createElement(View, { style: [styles$b.item, styles$b.left, itemHasCenterStyle] }, Left),
  !!Center && /*#__PURE__*/React.createElement(View, { style: [styles$b.item, styles$b.center] }, Center), /*#__PURE__*/
  React.createElement(View, { style: [styles$b.item, styles$b.right, itemHasCenterStyle] }, Right)));
}
var HeaderBase = /*#__PURE__*/memo(HeaderBaseComponent);

var styles$a = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center' },

  input: {
    padding: 10,
    flex: 1 },

  rightItem: {
    flexDirection: 'row',
    alignItems: 'center' } });

function defaultRenderProps() {
  return null;
}
function InputBaseComponent(_a) {
  var containerStyle = _a.containerStyle,inputStyle = _a.inputStyle,_b = _a.Left,Left = _b === void 0 ? defaultRenderProps : _b,_c = _a.Right,Right = _c === void 0 ? defaultRenderProps : _c,inputRef = _a.inputRef,otherProps = __rest(_a, ["containerStyle", "inputStyle", "Left", "Right", "inputRef"]);
  return /*#__PURE__*/React.createElement(View, { style: [styles$a.container, StyleSheet.flatten(containerStyle)] },
  Left, /*#__PURE__*/
  React.createElement(TextInput, __assign({}, otherProps, { ref: inputRef, underlineColorAndroid: "transparent", style: [styles$a.input, StyleSheet.flatten(inputStyle)] })), /*#__PURE__*/
  React.createElement(View, { style: styles$a.rightItem }, Right));
}
var InputBase = /*#__PURE__*/memo(InputBaseComponent);

var styles$9 = StyleSheet.create({
  br4: {
    borderRadius: 4 } });


function getBorderRadiusStyle(borderRadius) {
  if (typeof borderRadius === 'number') {
    return {
      borderRadius: borderRadius };

  }
  switch (borderRadius) {
    case 'round':
      return styles$9.br4;
    case 'pill':
      return tachyonsStyles.brPill;
    default:
      return {};}

}

var styles$8 = StyleSheet.create({
  bw1: {
    borderWidth: 1 },

  bw2: {
    borderWidth: 2 },

  bw3: {
    borderWidth: 4 } });


function getBorderWidthStyle(borderWidth) {
  switch (borderWidth) {
    case 1:
      return styles$8.bw1;
    case 2:
      return styles$8.bw2;
    case 3:
    default:
      return styles$8.bw3;}

}

var styles$7 = StyleSheet.create({
  input: {
    paddingVertical: isIOS ? 10 : 7 },

  hasLeft: {
    paddingLeft: 0 },

  hasRight: {
    paddingRight: 0 },

  right: {
    marginRight: -7 },

  iconX: {
    justifyContent: 'center',
    alignItems: 'center' },

  borderBottom: {
    borderWidth: 0,
    borderBottomWidth: 1 } });

function Input(_a) {
  var onChangeText = _a.onChangeText,onFocusText = _a.onFocusText,onFocus = _a.onFocus,_b = _a.onClearText,onClearText = _b === void 0 ? function () {} : _b,containerStyle = _a.containerStyle,inputStyle = _a.inputStyle,_c = _a.clearButtonMode,clearButtonMode = _c === void 0 ? 'never' : _c,ClearButtonModeComponent = _a.ClearButtonModeComponent,_d = _a.placeholder,placeholder = _d === void 0 ? '' : _d,_e = _a.borderRadius,borderRadius = _e === void 0 ? 'round' : _e,_f = _a.borderWidth,borderWidth = _f === void 0 ? 1 : _f,_g = _a.color,color = _g === void 0 ? 'dark1' : _g,_h = _a.backgroundColor,backgroundColor = _h === void 0 ? 'light' : _h,_j = _a.borderColor,borderColor = _j === void 0 ? 'dark4' : _j,Left = _a.Left,Right = _a.Right,otherProps = __rest(_a, ["onChangeText", "onFocusText", "onFocus", "onClearText", "containerStyle", "inputStyle", "clearButtonMode", "ClearButtonModeComponent", "placeholder", "borderRadius", "borderWidth", "color", "backgroundColor", "borderColor", "Left", "Right"]);
  var _k = useState(''),value = _k[0],setValue = _k[1];
  var _l = useTheme(),colors = _l.colors,sizes = _l.sizes;
  var inputHasLeftStyle = !!Left ? styles$7.hasLeft : {};
  var inputHasRightStyle = !!Left ? styles$7.hasRight : {};
  var _handleChangeText = function (text) {
    setValue(text);
    onChangeText === null || onChangeText === void 0 ? void 0 : onChangeText(text);
  };
  var _handleClearText = function () {
    setValue('');
    onClearText('');
  };
  var _handleFocus = function (event) {
    onFocusText === null || onFocusText === void 0 ? void 0 : onFocusText(event.nativeEvent.text);
    onFocus === null || onFocus === void 0 ? void 0 : onFocus(event);
  };
  var ClearButton = /*#__PURE__*/React.createElement(BaseButton, { style: [tachyonsStyles.justifyCenter, tachyonsStyles.itemsCenter], onPress: _handleClearText }, ClearButtonModeComponent !== null && ClearButtonModeComponent !== void 0 ? ClearButtonModeComponent : /*#__PURE__*/React.createElement(Feather, { name: "x", color: "dark3", size: sizes.base * 1.4, style: [tachyonsStyles.pv2, tachyonsStyles.pr2] }));
  var checkClearMode = {
    always: ClearButton,
    'while-editing': ClearButton,
    'unless-editing': null,
    never: null };

  return /*#__PURE__*/React.createElement(InputBase, __assign({}, otherProps, { value: value, onChangeText: _handleChangeText, onFocus: _handleFocus, placeholder: placeholder, Left: /*#__PURE__*/React.createElement(View, null, Left), Right: [/*#__PURE__*/React.createElement(View, { key: "item1" }, Right), /*#__PURE__*/React.createElement(View, { key: "item2" }, !!value && checkClearMode[clearButtonMode])], inputStyle: [styles$7.input, inputHasLeftStyle, inputHasRightStyle, { color: colors[color] }, inputStyle], containerStyle: [
    getBorderRadiusStyle(borderRadius),
    getBorderWidthStyle(borderWidth),
    { backgroundColor: colors[backgroundColor], borderColor: colors[borderColor] },
    containerStyle] }));

}

var styles$6 = StyleSheet.create({
  container: {
    flexWrap: 'wrap' },

  touchable: {
    justifyContent: 'center',
    alignItems: 'center',
    overflow: 'hidden' },

  inner: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center' },

  extraSmall: {
    paddingVertical: 2,
    paddingHorizontal: 6 },

  small: {
    paddingVertical: 6,
    paddingHorizontal: 12 },

  medium: {
    paddingVertical: 13,
    paddingHorizontal: 16 },

  large: {
    paddingVertical: 15,
    paddingHorizontal: 20 } });

function _getSizeStyle(size) {
  switch (size) {
    case 'extra-small':
      return styles$6.extraSmall;
    case 'small':
      return styles$6.small;
    case 'medium':
      return styles$6.medium;
    case 'large':
      return styles$6.large;
    default:
      return {};}

}
function _getFontSizeStyle(size) {
  if (size === 'large' || size === 'medium') {
    return tailwindStyles['text-sm'];
  }
  if (size === 'extra-small') {
    return tailwindStyles['text-xs'];
  }
  return {};
}
var TouchableOpacityWithStyles = withViewStyles(TouchableOpacity);
function ButtonComponent(_a) {
  var children = _a.children,_b = _a.borderRadius,borderRadius = _b === void 0 ? 'default' : _b,_c = _a.size,size = _c === void 0 ? 'medium' : _c,_d = _a.block,block = _d === void 0 ? false : _d,_e = _a.activeOpacity,activeOpacity = _e === void 0 ? 0.8 : _e,_f = _a.backgroundColor,backgroundColor = _f === void 0 ? 'primary' : _f,_g = _a.color,color = _g === void 0 ? 'light' : _g,_h = _a.borderWidth,borderWidth = _h === void 0 ? 1 : _h,_j = _a.borderColor,borderColor = _j === void 0 ? 'transparent' : _j,_k = _a.loading,loading = _k === void 0 ? false : _k,_l = _a.loadingColor,loadingColor = _l === void 0 ? 'light' : _l,LoadingComponent = _a.LoadingComponent,style = _a.style,otherProps = __rest(_a, ["children", "borderRadius", "size", "block", "activeOpacity", "backgroundColor", "color", "borderWidth", "borderColor", "loading", "loadingColor", "LoadingComponent", "style"]);
  var colors = useTheme().colors;
  var borderStyle = getBorderRadiusStyle(borderRadius);
  var sizeStyle = _getSizeStyle(size);
  var fontSizeStyle = _getFontSizeStyle(size);
  var blockStyle = block ? {} : tailwindStyles['flex-wrap'];
  var borderWidthStyle = getBorderWidthStyle(borderWidth);
  return /*#__PURE__*/React.createElement(View, { style: [blockStyle] }, /*#__PURE__*/
  React.createElement(TouchableOpacityWithStyles, __assign({}, otherProps, { style: [
    styles$6.touchable,
    borderStyle,
    sizeStyle,
    borderWidthStyle,
    { backgroundColor: colors[backgroundColor], borderColor: colors[borderColor] },
    style],
    activeOpacity: activeOpacity }), /*#__PURE__*/
  React.createElement(View, { style: styles$6.inner }, loading ? LoadingComponent !== null && LoadingComponent !== void 0 ? LoadingComponent : /*#__PURE__*/React.createElement(ActivityIndicator, { size: "small", color: colors[loadingColor] }) : typeof children === 'string' ? /*#__PURE__*/React.createElement(Text, { color: color, tachyons: "tc", style: fontSizeStyle }, children) : children)));
}
var Button = /*#__PURE__*/memo(ButtonComponent);

var styles$5 = StyleSheet.create({
  container: {
    position: 'relative',
    overflow: 'hidden' } });

var DEFAULT_IMAGE = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAA9JREFUeNpifvfuHUCAAQAFpALOO255kgAAAABJRU5ErkJggg==';
var ImageComponent = function (_a) {
  var _b = _a.uri,uri = _b === void 0 ? DEFAULT_IMAGE : _b,_c = _a.preview,preview = _c === void 0 ? DEFAULT_IMAGE : _c,_d = _a.containerStyle,containerStyle = _d === void 0 ? {} : _d,_e = _a.percentRatio,percentRatio = _e === void 0 ? '' : _e,_f = _a.borderRadius,borderRadius = _f === void 0 ? 'default' : _f,_g = _a.tachyons,tachyons = _g === void 0 ? [] : _g,onLoadEnd = _a.onLoadEnd,_h = _a.resizeMode,resizeMode = _h === void 0 ? 'cover' : _h,width = _a.width,height = _a.height,_j = _a.loading,loading = _j === void 0 ? false : _j,rest = __rest(_a, ["uri", "preview", "containerStyle", "percentRatio", "borderRadius", "tachyons", "onLoadEnd", "resizeMode", "width", "height", "loading"]);
  var _k = useState(percentRatio || '75%'),percentRatioState = _k[0],setPercentRatioState = _k[1];
  var _l = useState(false),isPreviewLoaded = _l[0],setPreviewLoaded = _l[1];
  var _m = useState(false),isUriLoaded = _m[0],setUriLoaded = _m[1];
  var req = useRef(-1);
  var checkHeight = path(['height'], containerStyle);
  var styled = useTheme().styled;
  useMount(function () {
    if (!!percentRatio) {
      setPercentRatioState(percentRatio);
    }
  });
  useUnmount(function () {
    req.current && cancelAnimationFrame(req.current);
  });
  var handleGetImageSizeSuccess = function (width, height) {
    setPercentRatioState(height / width * 100 + "%");
  };
  var handleGetImageSizeFailed = function () {
    console.log('Image getSize failed');
    req.current && cancelAnimationFrame(req.current);
  };
  var handleLoadEnd = function () {
    if (!percentRatio) {
      req.current = requestAnimationFrame(function () {
        Image$1.getSize(uri, handleGetImageSizeSuccess, handleGetImageSizeFailed);
      });
    }
    setUriLoaded(true);
    onLoadEnd === null || onLoadEnd === void 0 ? void 0 : onLoadEnd();
  };
  var handlePreviewLoadEnd = function () {
    setPreviewLoaded(true);
  };
  return /*#__PURE__*/React.createElement(View, { tachyons: tachyons, style: [
    styles$5.container,
    getBorderRadiusStyle(borderRadius),
    styled.bgGray2,
    containerStyle,
    !!width ? { width: width } : {},
    !!height ? { height: height } : {}] },

  !checkHeight && /*#__PURE__*/React.createElement(View, { style: { paddingTop: percentRatioState }, tachyons: ['flex', 'itemsCenter', 'justifyCenter'] }),
  !loading && !!preview && !isUriLoaded && /*#__PURE__*/React.createElement(Image$1, { source: { uri: preview }, resizeMode: resizeMode, style: [tachyonsStyles.absolute, tachyonsStyles.absoluteFill, tachyonsStyles.z1], onLoadEnd: handlePreviewLoadEnd, blurRadius: 1 }),
  !loading && isPreviewLoaded && !!uri && /*#__PURE__*/React.createElement(Image$1, __assign({}, rest, { resizeMode: resizeMode, source: { uri: uri }, style: [tachyonsStyles.absolute, tachyonsStyles.absoluteFill, tachyonsStyles.z2], onLoadEnd: handleLoadEnd })),
  loading && /*#__PURE__*/React.createElement(View, { tachyons: ['absolute', 'absoluteFill', 'itemsCenter', 'justifyCenter'] }, /*#__PURE__*/
  React.createElement(Feather, { name: "image", size: 50, color: "gray1" })));
};
var Image = /*#__PURE__*/memo(ImageComponent);

var styles$4 = StyleSheet.create({
  header: {
    overflow: 'hidden' },

  body: {},
  shadow: {
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 8 },
    shadowOpacity: 0.08,
    shadowRadius: 5,
    elevation: 3 },

  divider: {
    borderBottomWidth: 1 } });

function CardComponent(_a) {
  var Header = _a.Header,Body = _a.Body,Footer = _a.Footer,_b = _a.borderColor,borderColor = _b === void 0 ? 'gray1' : _b,borderWidth = _a.borderWidth,shadow = _a.shadow,_c = _a.backgroundColor,backgroundColor = _c === void 0 ? 'light' : _c,_d = _a.useDivider,useDivider = _d === void 0 ? false : _d,_e = _a.dividerColor,dividerColor = _e === void 0 ? 'gray2' : _e,_f = _a.style,style = _f === void 0 ? {} : _f,_g = _a.borderRadius,borderRadius = _g === void 0 ? 0 : _g,otherProps = __rest(_a, ["Header", "Body", "Footer", "borderColor", "borderWidth", "shadow", "backgroundColor", "useDivider", "dividerColor", "style", "borderRadius"]);
  var colors = useTheme().colors;
  var borderWidthStyle = !!borderWidth ? getBorderWidthStyle(borderWidth) : {};
  var shadowStyle = shadow ? styles$4.shadow : {};
  var dividerStyle = useDivider ? [styles$4.divider, { borderColor: colors[dividerColor] }] : {};
  var inlineStyle = {
    borderColor: colors[borderColor],
    backgroundColor: colors[backgroundColor],
    borderRadius: borderRadius };

  return /*#__PURE__*/React.createElement(View, __assign({}, otherProps, { style: [borderWidthStyle, shadowStyle, inlineStyle, style] }),
  !!Header && /*#__PURE__*/React.createElement(View, { style: [styles$4.header, !!Body || !!Footer ? dividerStyle : {}] }, Header),
  !!Body && /*#__PURE__*/React.createElement(View, { style: !!Footer ? dividerStyle : {} }, Body),
  !!Footer && /*#__PURE__*/React.createElement(View, null, Footer));
}
var Card = /*#__PURE__*/memo(withTailwind(CardComponent));

var Icons = {
  FontAwesome: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    return /*#__PURE__*/React.createElement(FontAwesome, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }),
  FontAwesome5: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    return /*#__PURE__*/React.createElement(FontAwesome5, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }),
  Feather: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    return /*#__PURE__*/React.createElement(Feather, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }),
  MaterialIcons: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    return /*#__PURE__*/React.createElement(MaterialIcons, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }),
  MaterialCommunityIcons: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    // @ts-ignore
    return /*#__PURE__*/React.createElement(MaterialCommunityIcons, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }),
  AntDesign: /*#__PURE__*/memo(function (_a) {
    var _b = _a.color,color = _b === void 0 ? 'dark2' : _b,colorNative = _a.colorNative,rest = __rest(_a, ["color", "colorNative"]);
    var colors = useTheme().colors;
    return /*#__PURE__*/React.createElement(AntDesign, __assign({}, rest, { color: !!colorNative ? colorNative : colors[color] }));
  }) };

var styles$3 = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    flexWrap: 'nowrap',
    overflow: 'hidden',
    marginVertical: 15 },

  textWrap: {
    position: 'relative',
    marginHorizontal: 15 },

  divider: {
    width: '100%',
    height: 1 } });

function TextDividerComponennt(_a) {
  var _b = _a.textStyle,textStyle = _b === void 0 ? {} : _b,_c = _a.containerStyle,containerStyle = _c === void 0 ? {} : _c,children = _a.children;
  var styled = useTheme().styled;
  return /*#__PURE__*/React.createElement(View$1, { style: [styles$3.container, StyleSheet.flatten(containerStyle)] }, /*#__PURE__*/
  React.createElement(View$1, { style: [styles$3.divider, styled.bgDark4] }), /*#__PURE__*/
  React.createElement(View$1, { style: styles$3.textWrap }, /*#__PURE__*/
  React.createElement(Text$1, { style: [tailwindStyles['text-sm'], styled.colorDark3, StyleSheet.flatten(textStyle)] }, children)), /*#__PURE__*/
  React.createElement(View$1, { style: [styles$3.divider, styled.bgDark4] }));
}
var TextDivider = /*#__PURE__*/memo(TextDividerComponennt);

var initialMeasure = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  pageX: 0,
  pageY: 0 };

var _req = 0;
function useMeasure(ref) {
  var _a = useState(initialMeasure),measure = _a[0],setMeasure = _a[1];
  var onMeasure = useCallback(function () {
    _req = requestAnimationFrame(function () {
      !!ref.current &&
      ref.current.measure(function (x, y, width, height, pageX, pageY) {
        setMeasure({ x: x, y: y, width: width, height: height, pageX: pageX, pageY: pageY });
      });
    });
  }, [ref]);
  useEffect(function () {
    return function () {
      !!_req && cancelAnimationFrame(_req);
    };
  }, []);
  return [measure, onMeasure];
}

function useToggle(initialState) {
  var _this = this;
  var _a = useState(initialState),state = _a[0],setState = _a[1];
  var onToggle = function () {return __awaiter(_this, void 0, void 0, function () {
      var nextState;
      return __generator(this, function (_a) {
        nextState = !state;
        setState(nextState);
        return [2 /*return*/, new Promise(function (resolve) {
          resolve(nextState);
        })];
      });
    });};
  return [state, onToggle];
}

function sleep(ms) {
  return new Promise(function (resolve) {
    var timeout = setTimeout(function () {
      resolve(undefined);
      clearTimeout(timeout);
    }, ms);
    return timeout;
  });
}

function getInlineStyle(placement, buttonMeasure, contentMeasure) {
  var top = buttonMeasure.pageY - contentMeasure.height;
  var bottom = buttonMeasure.pageY + buttonMeasure.height;
  var leftInner = buttonMeasure.pageX;
  var rightInner = buttonMeasure.pageX - contentMeasure.width + buttonMeasure.width;
  var leftOuter = buttonMeasure.pageX - contentMeasure.width;
  var rightOuter = contentMeasure.width + buttonMeasure.width;
  var center = buttonMeasure.pageX - (contentMeasure.width - buttonMeasure.width) / 2;
  var middle = buttonMeasure.pageY - (contentMeasure.height - buttonMeasure.height) / 2;
  switch (placement) {
    case 'topLeft':
      return {
        top: top,
        left: leftInner };

    case 'topRight':
      return {
        top: top,
        left: rightInner };

    case 'rightTop':
      return {
        top: top,
        left: rightOuter };

    case 'right':
      return {
        top: middle,
        left: rightOuter };

    case 'rightBottom':
      return {
        top: bottom,
        left: rightOuter };

    case 'bottomLeft':
      return {
        top: bottom,
        left: leftInner };

    case 'bottom':
      return {
        top: bottom,
        left: center };

    case 'bottomRight':
      return {
        top: bottom,
        left: rightInner };

    case 'leftTop':
      return {
        top: top,
        left: leftOuter };

    case 'left':
      return {
        top: middle,
        left: leftOuter };

    case 'leftBottom':
      return {
        top: bottom,
        left: leftOuter };

    case 'top':
    default:
      return {
        top: top,
        left: center };}


}
function useSwitchAnimation(_a) {
  var _this = this;
  var placement = _a.placement,_b = _a.duration,duration = _b === void 0 ? 0 : _b;
  var buttonRef = useRef(null);
  var contentRef = useRef(null);
  var _c = useToggle(false),isVisible = _c[0],onVisible = _c[1];
  var _d = useToggle(false),isAnimation = _d[0],onAnim = _d[1];
  var _e = useMeasure(buttonRef),buttonMeasure = _e[0],onButtonMeasure = _e[1];
  var _f = useMeasure(contentRef),contentMeasure = _f[0],onContentMeasure = _f[1];
  var contentStyle = getInlineStyle(placement, buttonMeasure, contentMeasure);
  var onOpen = function () {return __awaiter(_this, void 0, void 0, function () {
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            onButtonMeasure();
            onContentMeasure();
            onVisible();
            return [4 /*yield*/, sleep(100)];
          case 1:
            _a.sent();
            onAnim();
            return [2 /*return*/];}

      });
    });};
  var onClose = function () {return __awaiter(_this, void 0, void 0, function () {
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            onAnim();
            return [4 /*yield*/, sleep(duration)];
          case 1:
            _a.sent();
            onVisible();
            return [2 /*return*/];}

      });
    });};
  return { buttonRef: buttonRef, contentRef: contentRef, contentStyle: contentStyle, isVisible: isVisible, isAnimation: isAnimation, onOpen: onOpen, onClose: onClose };
}

var styles$2 = StyleSheet.create({
  popoverOuter: {
    position: 'absolute',
    zIndex: 10 },

  underlay: __assign(__assign({}, StyleSheet.absoluteFillObject), { zIndex: 9 }),
  popover: {
    paddingTop: 10 },

  popoverInner: {},
  shadow: {
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 3 },
    shadowOpacity: 0.12,
    shadowRadius: 5,
    elevation: 3 } });

// const AnimatedView = posed.View({
//   open: { opacity: 1, scale: 1, transition: { duration: 100 } },
//   closed: { opacity: 0, scale: 0.4, transition: { duration: 100 } },
// });
function Popover(_a) {
  var children = _a.children,Content = _a.Content,_b = _a.style,style = _b === void 0 ? {} : _b,_c = _a.placement,placement = _c === void 0 ? 'top' : _c,_d = _a.appearance,appearance = _d === void 0 ? 'light' : _d;
  var styled = useTheme().styled;
  var _e = useSwitchAnimation({
    placement: placement,
    duration: 100 }),
  buttonRef = _e.buttonRef,contentRef = _e.contentRef,contentStyle = _e.contentStyle,isVisible = _e.isVisible,onClose = _e.onClose,onOpen = _e.onOpen;
  var popoverInnerStyle = [styles$2.popoverInner, styles$2.shadow, appearance === 'light' ? styled.bgLight : styled.bgDark1];
  var popoverContentStyle = [styles$2.popoverOuter, contentStyle];
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/
  React.createElement(Modal$1, { visible: isVisible, animationType: "fade", transparent: true, onRequestClose: onClose }, /*#__PURE__*/
  React.createElement(View, { ref: contentRef, style: popoverContentStyle }, /*#__PURE__*/
  React.createElement(View, { style: popoverInnerStyle }, Content)), /*#__PURE__*/
  React.createElement(TouchableOpacity, { activeOpacity: 1, onPress: onClose, style: styles$2.underlay })), /*#__PURE__*/
  React.createElement(TouchableOpacity, { ref: buttonRef, onPress: onOpen, style: style }, children));
}

function useSelectList(_a) {
  var _b = _a.inputResult,inputResult = _b === void 0 ? [] : _b,_c = _a.multiple,multiple = _c === void 0 ? false : _c,_d = _a.onResultCallback,onResultCallback = _d === void 0 ? function () {} : _d;
  var _e = useState([]),outputResult = _e[0],setResult = _e[1];
  var inputResultPrevRef = useRef([]);
  var onResultCallbackRef = useRef(onResultCallback);
  var setInputResult = useCallback(function () {
    var _a;
    // gi tr default nu khng phi dng multiple th ch c 1 phn t u tin
    if (!equals(inputResultPrevRef.current, inputResult)) {
      var nextOutputResult = multiple ? inputResult : [inputResult[0]];
      setResult(nextOutputResult);
      (_a = onResultCallbackRef.current) === null || _a === void 0 ? void 0 : _a.call(onResultCallbackRef, nextOutputResult);
    }
    inputResultPrevRef.current = inputResult;
  }, [inputResult, multiple]);
  useEffect(function () {
    setInputResult();
  }, [setInputResult]);
  var onSelect = useCallback(function (item) {
    var removeItem = outputResult.filter(function (_item) {return !equals(_item, item);});
    var addItem = multiple ? __spreadArray(__spreadArray([], outputResult), [item]) : [item];
    var nextOutputResult = includes(item, outputResult) ? removeItem : addItem;
    setResult(nextOutputResult);
    onResultCallback === null || onResultCallback === void 0 ? void 0 : onResultCallback(nextOutputResult);
  },
  // eslint-disable
  [outputResult, multiple, onResultCallback]);
  var isSelected = function (item) {return includes(item, outputResult);};
  var getNumberSelected = function (item) {return outputResult.findIndex(function (_item) {return equals(_item, item);}) + 1;};
  return { outputResult: outputResult, isSelected: isSelected, getNumberSelected: getNumberSelected, onSelect: onSelect, onPrevious: setInputResult };
}

function SelectList(_a) {
  var renderItem = _a.renderItem,onResult = _a.onResult,multiple = _a.multiple,defaultResult = _a.defaultResult,flatListProps = __rest(_a, ["renderItem", "onResult", "multiple", "defaultResult"]);
  var _b = useSelectList({
    inputResult: defaultResult,
    multiple: multiple,
    onResultCallback: onResult }),
  isSelected = _b.isSelected,onSelect = _b.onSelect;
  var _renderItem = function (_a) {
    var item = _a.item,index = _a.index,separators = _a.separators;
    var selected = isSelected(item);
    return renderItem({ item: item, index: index, separators: separators, selected: selected, onSelect: onSelect });
  };
  return /*#__PURE__*/React.createElement(FlatList$1, __assign({}, flatListProps, { renderItem: _renderItem }));
}
SelectList.defaultProps = {
  multiple: true,
  defaultResult: [] };

var styles$1 = StyleSheet.create({
  container: {
    height: 1 } });


var Divider = withTailwind(function (_a) {
  var _b = _a.color,color = _b === void 0 ? 'gray2' : _b,style = _a.style;
  var colors = useTheme().colors;
  return /*#__PURE__*/React.createElement(View, { style: [styles$1.container, { backgroundColor: colors[color] }, style] });
});

var Modal = function (_a) {
  var children = _a.children,_b = _a.visible,visible = _b === void 0 ? false : _b,style = _a.style,_c = _a.animationType,animationType = _c === void 0 ? 'none' : _c,_d = _a.animationDuration,animationDuration = _d === void 0 ? 400 : _d,_e = _a.animmationEasing,animmationEasing = _e === void 0 ? Easing.bezier(0.46, 0.79, 0.54, 0.98) : _e,otherProps = __rest(_a, ["children", "visible", "style", "animationType", "animationDuration", "animmationEasing"]);
  var _f = usePortalAnimation({ visible: visible, animationType: animationType, animationDuration: animationDuration, animmationEasing: animmationEasing }),visibleState = _f.visibleState,animationStyle = _f.animationStyle;
  return /*#__PURE__*/React.createElement(ModalBase, __assign({ visible: visibleState }, otherProps), /*#__PURE__*/
  React.createElement(Animated.View, { style: [tailwindStyles['flex-1'], animationStyle] }, /*#__PURE__*/
  React.createElement(View, { flex: true, style: style }, children)));
};

function FlatList(_a) {
  var _b = _a.useContainer,useContainer = _b === void 0 ? false : _b,renderItem = _a.renderItem,_c = _a.numColumns,numColumns = _c === void 0 ? 1 : _c,_d = _a.numGap,numGap = _d === void 0 ? 16 : _d,data = _a.data,navigation = _a.navigation,contentContainerStyle = _a.contentContainerStyle,columnWrapperStyle = _a.columnWrapperStyle,horizontal = _a.horizontal,ListHeaderComponent = _a.ListHeaderComponent,ListEmptyComponent = _a.ListEmptyComponent,refreshing = _a.refreshing,onRefresh = _a.onRefresh,onScroll = _a.onScroll,_e = _a.showsVerticalScrollIndicator,showsVerticalScrollIndicator = _e === void 0 ? false : _e,_f = _a.showsHorizontalScrollIndicator,showsHorizontalScrollIndicator = _f === void 0 ? false : _f,rest = __rest(_a, ["useContainer", "renderItem", "numColumns", "numGap", "data", "navigation", "contentContainerStyle", "columnWrapperStyle", "horizontal", "ListHeaderComponent", "ListEmptyComponent", "refreshing", "onRefresh", "onScroll", "showsVerticalScrollIndicator", "showsHorizontalScrollIndicator"]);
  var sizes = useTheme().sizes;
  var _g = useState(false),stateRefreshing = _g[0],setStateRefreshing = _g[1];
  var isTopRef = useRef(false);
  var flatListRef = useRef(null);
  var contentOffsetYRef = useRef(0);
  var ItemWrap = useContainer && numColumns === 1 ? Container : View;
  var ListComponentWrap = useContainer && isIOS ? Container : numColumns > 1 ? View : Container;
  var itemWrapStyle = numColumns > 1 ? { width: 100 / numColumns + "%" } : {};
  var listComponentWrapStyle = isIOS ? {} : numColumns > 1 ? { maxWidth: sizes.container } : {};
  var gapItemStyle = numColumns > 1 ? { paddingHorizontal: numGap / 2 } : {};
  var columnWrapperMaxWidth = sizes.container + (numColumns > 1 ? numGap : 0);
  useMount(function () {
    setStateRefreshing(refreshing);
    !!navigation &&
    navigation.setParams({
      onScrollToTop: function () {
        var _a;
        if (isTopRef.current && contentOffsetYRef.current === 0) {
          setStateRefreshing(true);
          var timeout_1 = setTimeout(function () {
            onRefresh === null || onRefresh === void 0 ? void 0 : onRefresh();
            setStateRefreshing(false);
            clearTimeout(timeout_1);
          }, 400);
          isTopRef.current = false;
        } else
        {
          (_a = flatListRef.current) === null || _a === void 0 ? void 0 : _a.scrollToOffset({ offset: 0, animated: true });
          isTopRef.current = true;
        }
      } });

  });
  var renderListComponent = function (Component) {
    return /*#__PURE__*/React.createElement(ListComponentWrap, __assign({}, horizontal ? { flex: 1 } : {}, { style: listComponentWrapStyle, innerTachyons: isIOS ? 'w100' : 'w100' }), Component);
  };
  /**
   * renderItemFixed
   * @param index - item index
   * @example
   * [item1] [item2] [item3]
   * [item4] [item5] [item6]
   * [item7] [fixed1] [fixed2]
   */
  var renderItemFixed = function (index) {
    return !!data &&
    data.length % numColumns !== 0 &&
    index === data.length - 1 &&
    range(0, numColumns - data.length % numColumns).map(function (item) {return /*#__PURE__*/React.createElement(View, { key: String(item), style: [itemWrapStyle, gapItemStyle] });});
  };
  var handleScroll = function (event) {
    onScroll === null || onScroll === void 0 ? void 0 : onScroll(event);
    contentOffsetYRef.current = event.nativeEvent.contentOffset.y;
  };
  return /*#__PURE__*/React.createElement(FlatList$1, __assign({}, rest, { data: data, ref: function (c) {
      flatListRef.current = c;
    }, numColumns: numColumns, horizontal: horizontal, refreshing: stateRefreshing, onRefresh: onRefresh, onScroll: handleScroll, showsVerticalScrollIndicator: showsVerticalScrollIndicator, showsHorizontalScrollIndicator: showsHorizontalScrollIndicator, ListHeaderComponent: renderListComponent(ListHeaderComponent),
    // ListFooterComponent={renderListComponent(ListFooterComponent)}
    ListEmptyComponent: renderListComponent(ListEmptyComponent), renderItem: function (info) {
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/
      React.createElement(ItemWrap, { style: [itemWrapStyle, gapItemStyle] }, renderItem === null || renderItem === void 0 ? void 0 : renderItem(info)),
      renderItemFixed(info.index));
    } }, useContainer && numColumns > 1 ?
  {
    contentContainerStyle: [tachyonsStyles.itemsCenter, contentContainerStyle],
    columnWrapperStyle: [{ maxWidth: columnWrapperMaxWidth }, columnWrapperStyle] } :

  {
    contentContainerStyle: contentContainerStyle,
    columnWrapperStyle: columnWrapperStyle }));

}

var OfflineNotice = function (_a) {
  var _b = _a.backgroundColor,backgroundColor = _b === void 0 ? 'dark1' : _b,_c = _a.children,children = _c === void 0 ? 'No Internet Connection' : _c,rest = __rest(_a, ["backgroundColor", "children"]);
  var _d = useState(true),isConnected = _d[0],setIsConnected = _d[1];
  useEffect(function () {
    var unsubscribe = NetInfo.addEventListener(function (state) {
      setIsConnected(!!state.isConnected);
    });
    return function () {
      unsubscribe();
    };
  }, []);
  if (isConnected) {
    return null;
  }
  return /*#__PURE__*/React.createElement(View, __assign({}, rest, { backgroundColor: backgroundColor }), typeof children === 'string' ? /*#__PURE__*/React.createElement(View, { tachyons: ['itemsCenter', 'justifyCenter', 'flexRow', 'pa2'] }, /*#__PURE__*/
  React.createElement(Feather, { name: "alert-triangle", color: "gray1", size: 18 }), /*#__PURE__*/
  React.createElement(Text, { color: "gray1", tachyons: ['tc', 'pl2'] }, children)) : children);
};

function checkInRange(source, from, to) {
  return source > from && source < to;
}
function toAlphabetLowerCase(souce) {
  return souce.
  normalize('NFD').
  replace(/[\u0300-\u036f]/g, '').
  replace(/[]/g, function (m) {return m === '' ? 'd' : 'D';}).
  toLowerCase();
}
function getInputFakePaddingVerticalValue(value) {
  var checkTextMultipeLine = /\n/g.test(value);
  var forIos = 7;
  var forAndroid = checkTextMultipeLine ? 5 : 10;
  return isIOS ? forIos : forAndroid;
}
function joinStringInArray(source) {
  return pipe(groupWith(function (x, y) {return type(x) === 'String' && type(y) === 'String';}), map(function (item) {return item.length === 1 ? item[0] : item.join('');}))(source);
}

var styles = StyleSheet.create({
  inputContainer: {
    position: 'relative' },

  input: {
    padding: 10,
    fontSize: 14,
    lineHeight: 20 },

  inputImportant: {
    color: 'transparent',
    backgroundColor: 'transparent' },

  inputFakeImportant: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    zIndex: -1 } });

function TextHighlight(_a) {
  var value = _a.value,_b = _a.mentionStyle,mentionStyle = _b === void 0 ? {} : _b,entityMap = _a.entityMap,_c = _a.containerStyle,containerStyle = _c === void 0 ? {} : _c,onPressHighlight = _a.onPressHighlight,_d = _a.onPressHighlightEnabled,onPressHighlightEnabled = _d === void 0 ? false : _d;
  return /*#__PURE__*/React.createElement(Text, { style: containerStyle }, joinStringInArray(value.split('').map(function (char, index) {
    var _a;
    var _loop_1 = function (i) {
      var condition = checkInRange(index, entityMap[i].range.offset - 1, entityMap[i].range.offset + entityMap[i].range.length) &&
      !!((_a = entityMap[i]) === null || _a === void 0 ? void 0 : _a.mentions) &&
      !isEmpty(entityMap[i].mentions);
      if (condition) {
        return { value: /*#__PURE__*/React.createElement(Text, { key: String(index), style: mentionStyle, onPress: onPressHighlightEnabled ? function () {var _a;return onPressHighlight((_a = entityMap[i]) === null || _a === void 0 ? void 0 : _a.mentions);} : function () {} }, char) };
      }
    };
    for (var i = 0; i < entityMap.length; i++) {
      var state_1 = _loop_1(i);
      if (typeof state_1 === "object")
      return state_1.value;
    }
    return char;
  })));
}
var typedMemo = memo;
var TextHighlight$1 = typedMemo(TextHighlight);

function ListUser(_a) {
  var _b = _a.users,users = _b === void 0 ? [] : _b,_c = _a.containerStyle,containerStyle = _c === void 0 ? {} : _c,_d = _a.readonly,readonly = _d === void 0 ? false : _d,renderUserItem = _a.renderUserItem,keyExtractor = _a.keyExtractor,hideUserMentioned = _a.hideUserMentioned,keyForMention = _a.keyForMention,search = _a.search,entityMap = _a.entityMap,onItemPress = _a.onItemPress;
  var getMentions = function () {
    return entityMap.map(function (item) {return item.mentions;});
  };
  return /*#__PURE__*/React.createElement(ScrollView, { style: containerStyle, keyboardShouldPersistTaps: "handled" }, !isEmpty(users) &&
  !readonly &&
  users.map(function (user, index) {
    var mentions = getMentions();
    var at = search.charAt(0);
    var searchNotAt = search.replace(new RegExp("^" + at, 'g'), '');
    var checkHideUserMentioned = hideUserMentioned && includes(user, mentions);
    var atOnly = at && !searchNotAt;
    var userHasSearch = toAlphabetLowerCase(user[keyForMention]).includes(toAlphabetLowerCase(searchNotAt) || '*');
    if (checkHideUserMentioned || !atOnly && !userHasSearch) {
      return null;
    }
    return /*#__PURE__*/React.createElement(TouchableOpacity, { key: keyExtractor(user, index), activeOpacity: 1, onPress: function () {
        onItemPress === null || onItemPress === void 0 ? void 0 : onItemPress(user);
      } }, renderUserItem === null || renderUserItem === void 0 ? void 0 : renderUserItem(user, index));
  }));
}

var TextInputMentions = /** @class */function (_super) {
  __extends(TextInputMentions, _super);
  function TextInputMentions(props) {
    var _this = _super.call(this, props) || this;
    _this.getCaretPosition = function () {
      var selection = _this.state.selection;
      return isIOS ? selection.start : selection.start + 1;
    };
    _this.handleDeleteText = function () {
      var _a = _this.state,value = _a.value,entityMap = _a.entityMap,textAddedLength = _a.textAddedLength;
      var caretPosition = _this.getCaretPosition();
      var isTextSelected = textAddedLength < -1;
      var WTF_androidFrom = isIOS ? 1 : -1;
      var WTF_androidTo = isIOS ? 0 : 2;
      // trng hp offset nm trong on ang bi en
      var checkOffsetInnerSelection = function (range) {return checkInRange(range.offset - WTF_androidFrom, caretPosition, caretPosition - textAddedLength);};
      // trng hp length nm trong on ang bi en
      var checkLengthInnerSelection = function (range) {
        return checkInRange(range.offset + range.length + WTF_androidTo, caretPosition, caretPosition - textAddedLength);
      };
      // trng hp on bi en nm trong range
      var checkSelectionInnerRange = function (range) {
        return range.offset - WTF_androidFrom < caretPosition && caretPosition - textAddedLength < range.offset + range.length + WTF_androidTo;
      };
      var newEntityMap = entityMap.filter(function (_a) {
        var range = _a.range;
        if (isTextSelected) {
          return !checkOffsetInnerSelection(range) && !checkLengthInnerSelection(range) && !checkSelectionInnerRange(range);
        }
        return !checkInRange(caretPosition, range.offset - WTF_androidFrom, range.offset + range.length + WTF_androidTo);
      });
      var entityMapDeleted = entityMap.filter(function (_a) {
        var range = _a.range;
        if (isTextSelected) {
          return checkOffsetInnerSelection(range) && checkLengthInnerSelection(range) && checkSelectionInnerRange(range);
        }
        return checkInRange(caretPosition, range.offset - WTF_androidFrom, range.offset + range.length + WTF_androidTo);
      })[0];
      _this.setState({
        entityMap: newEntityMap });

      if (!!(entityMapDeleted === null || entityMapDeleted === void 0 ? void 0 : entityMapDeleted.range)) {
        var range_1 = entityMapDeleted.range;
        var newValue = value.
        split('').
        filter(function (_, index) {return !checkInRange(index, range_1.offset - 1, range_1.offset + range_1.length);}).
        join('');
        _this.setState({
          value: newValue,
          textAddedLength: -(range_1.length + 1) });

      }
    };
    _this.handleUpdateRangeOffset = function () {
      var _a = _this.state,value = _a.value,entityMap = _a.entityMap,textAddedLength = _a.textAddedLength;
      var caretPosition = _this.getCaretPosition();
      if (caretPosition >= value.length && isEmpty(entityMap)) {
        return;
      }
      _this.setState({
        entityMap: entityMap.map(function (_a) {
          var range = _a.range,mentions = _a.mentions;
          var newRange = caretPosition <= range.offset + 1 ? __assign(__assign({}, range), { offset: range.offset + textAddedLength }) : range;
          return {
            mentions: mentions,
            range: newRange };

        }) });

    };
    _this.setValueAsync = function (value) {
      return new Promise(function (resolve) {
        _this.setState({ value: value }, function () {return resolve(undefined);});
      });
    };
    _this.setTextAddedLengthAsync = function (value, prevValue) {
      return new Promise(function (resolve) {
        _this.setState({ textAddedLength: value.length - prevValue.length }, function () {return resolve(undefined);});
      });
    };
    _this.handleAddTextInnerRange = function () {
      var caretPosition = _this.getCaretPosition();
      _this.setState(function (prevState) {return {
          entityMap: prevState.entityMap.filter(function (_a) {
            var range = _a.range;
            return !checkInRange(caretPosition, range.offset, range.offset + range.length);
          }) };
      });
    };
    _this.handleSearch = function () {
      var withChar = _this.props.withChar;
      var value = _this.state.value;
      var caretPosition = _this.getCaretPosition();
      var valueHasAt = value.includes(withChar);
      var reg = new RegExp(".*(?=" + withChar + ")", 'g');
      var searchValue = valueHasAt ?
      value.
      replace(/\n/g, '').
      substring(0, caretPosition).
      replace(reg, '') :
      '';
      searchValue = searchValue.charAt(1) === ' ' ? '' : searchValue;
      _this.setState({
        search: searchValue,
        atPosition: value.substring(0, caretPosition).length });

    };
    _this.handleChangeText = function (text) {return __awaiter(_this, void 0, void 0, function () {
        var prevValue;
        return __generator(this, function (_a) {
          switch (_a.label) {
            case 0:
              prevValue = this.state.value;
              return [4 /*yield*/, this.setValueAsync(text)];
            case 1:
              _a.sent();
              return [4 /*yield*/, this.setTextAddedLengthAsync(text, prevValue)];
            case 2:
              _a.sent();
              if (text.length <= prevValue.length) {
                this.handleDeleteText();
              } else
              {
                this.handleAddTextInnerRange();
              }
              this.handleUpdateRangeOffset();
              this.handleSearch();
              return [2 /*return*/];}

        });
      });};
    _this.insertNewItemEntityMap = function (user) {
      var keyForMention = _this.props.keyForMention;
      var _a = _this.state,search = _a.search,entityMap = _a.entityMap,atPosition = _a.atPosition;
      var newItem = {
        mentions: user,
        range: {
          offset: atPosition - search.length,
          length: user[keyForMention].length } };


      _this.setState({
        entityMap: insert(atPosition, newItem, entityMap),
        search: '' });

    };
    _this.handleUserItemPress = function (user) {
      var keyForMention = _this.props.keyForMention;
      var _a = _this.state,value = _a.value,entityMap = _a.entityMap,search = _a.search,atPosition = _a.atPosition;
      var newEntityMap = entityMap.map(function (_a) {
        var range = _a.range,mentions = _a.mentions;
        var condition = atPosition < value.length && atPosition <= range.offset + 1;
        var newRange = condition ? __assign(__assign({}, range), { offset: range.offset + user[keyForMention].length - search.length }) : range;
        return {
          mentions: mentions,
          range: newRange };

      });
      var newValue = insert(atPosition - search.length, user[keyForMention], (value.substr(0, atPosition - search.length) + value.substr(atPosition)).split('')).join('');
      _this.setState({
        value: newValue,
        entityMap: newEntityMap,
        textAddedLength: user[keyForMention].length },
      function () {
        _this.insertNewItemEntityMap(user);
      });
    };
    _this.handleSelectionChange = function (event) {
      _this.setState({
        selection: event.nativeEvent.selection });

    };
    _this.renderInput = function () {
      var _a = _this.props,readonly = _a.readonly,inputStyle = _a.inputStyle,inputRef = _a.inputRef,placeholder = _a.placeholder,placeholderTextColor = _a.placeholderTextColor;
      var value = _this.state.value;
      if (readonly) {
        return null;
      }
      return /*#__PURE__*/React.createElement(TextInput, { ref: inputRef, style: [styles.input, inputStyle, styles.inputImportant], multiline: true, value: value, placeholder: placeholder, placeholderTextColor: placeholderTextColor, onChangeText: _this.handleChangeText, onSelectionChange: _this.handleSelectionChange, autoCapitalize: "none", autoCompleteType: "off", autoCorrect: false });
    };
    _this.state = {
      value: props.value,
      prevValue: '',
      textAddedLength: 0,
      entityMap: props.entityMap,
      selection: { start: 0, end: 0 },
      search: '',
      atPosition: -1 };

    return _this;
  }
  TextInputMentions.prototype.componentDidUpdate = function (prevProps, prevState) {
    var _a = this.props,onChange = _a.onChange,valueProp = _a.value,entityMapProp = _a.entityMap;
    var _b = this.state,value = _b.value,entityMap = _b.entityMap;
    if (prevProps.value !== valueProp) {
      this.setState({
        value: valueProp });

    }
    if (!equals(prevProps.entityMap, entityMapProp)) {
      this.setState({
        entityMap: entityMapProp });

    }
    if (prevState.value !== value || !equals(prevState.entityMap, entityMap)) {
      onChange({ value: value, entityMap: entityMap });
    }
  };
  TextInputMentions.prototype.render = function () {
    var _a = this.props,users = _a.users,inputStyle = _a.inputStyle,containerStyle = _a.containerStyle,inputContainerStyle = _a.inputContainerStyle,userContainerStyle = _a.userContainerStyle,readonly = _a.readonly,mentionStyle = _a.mentionStyle,hideUserMentioned = _a.hideUserMentioned,keyForMention = _a.keyForMention,keyExtractor = _a.keyExtractor,renderUserItem = _a.renderUserItem,onPressMention = _a.onPressMention;
    var _b = this.state,value = _b.value,entityMap = _b.entityMap,search = _b.search;
    return /*#__PURE__*/React.createElement(View, { style: containerStyle }, /*#__PURE__*/
    React.createElement(ListUser, { entityMap: entityMap, users: users, readonly: readonly, keyForMention: keyForMention, keyExtractor: keyExtractor, renderUserItem: renderUserItem, hideUserMentioned: hideUserMentioned, search: search, containerStyle: userContainerStyle, onItemPress: this.handleUserItemPress }), /*#__PURE__*/
    React.createElement(View, { style: [styles.inputContainer, inputContainerStyle] },
    this.renderInput(), /*#__PURE__*/
    React.createElement(TextHighlight$1, { onPressHighlightEnabled: readonly, value: value, entityMap: entityMap, mentionStyle: mentionStyle, containerStyle: [
      styles.input,
      inputStyle,
      readonly ? {} : styles.inputFakeImportant,
      readonly ? {} : { paddingVertical: getInputFakePaddingVerticalValue(value) }],
      onPressHighlight: onPressMention })));
  };
  TextInputMentions.defaultProps = {
    users: [],
    renderUserItem: function () {return null;},
    keyExtractor: function (_, index) {return String(index);},
    hideUserMentioned: true,
    onChange: function () {},
    onPressMention: function () {},
    readonly: false,
    withChar: '@',
    value: '',
    entityMap: [],
    containerStyle: {},
    inputContainerStyle: {},
    userContainerStyle: {},
    inputStyle: {},
    mentionStyle: {
      backgroundColor: '#dce6f8' },

    keyForMention: 'name',
    inputRef: null,
    placeholder: '',
    placeholderTextColor: '#888' };

  return TextInputMentions;
}(PureComponent);

var NavigationSuspense = /** @class */function (_super) {
  __extends(NavigationSuspense, _super);
  function NavigationSuspense(props) {
    var _this = _super.call(this, props) || this;
    _this._handleLoaded = function () {
      _this.setState({
        isLoaded: true });

    };
    _this.state = {
      isLoaded: false };

    _this._req = requestAnimationFrame(_this._handleLoaded);
    return _this;
  }
  NavigationSuspense.prototype.componentWillUnmount = function () {
    cancelAnimationFrame(this._req);
  };
  NavigationSuspense.prototype.render = function () {
    var _a = this.props,children = _a.children,fallback = _a.fallback;
    var isLoaded = this.state.isLoaded;
    return isLoaded ? children : fallback;
  };
  NavigationSuspense.defaultProps = {
    fallback: null };

  return NavigationSuspense;
}(PureComponent);

var handleAnchorElement = function (baseUrl) {return "\nconst anchorClicks = [];\n\nconst getAttributes = el => {\n  return el.getAttributeNames().reduce((obj, name) => ({\n    ...obj,\n    [name]: el.getAttribute(name)\n  }), {});\n}\n\nif (!window.getEventListeners) {\n  window.getEventListeners = () => []\n}\n\nconst handleTouch = event => {\n  const anchorEl = event.target.tagName === 'A' ? event.target : event.target.closest('a');\n  const isAnchorListenerEmpty = !window.getEventListeners(anchorEl, 'click').length;\n  const isMyListener = !isAnchorListenerEmpty && window.getEventListeners(anchorEl, 'click') && window.getEventListeners(anchorEl, 'click')[0] && window.getEventListeners(anchorEl, 'click')[0].listener && window.getEventListeners(anchorEl, 'click')[0].listener.toString() && window.getEventListeners(anchorEl, 'click')[0].listener.toString().includes('!!anchorEl.href &&');\n\n  if (!anchorClicks.includes(anchorEl) && (isAnchorListenerEmpty || isMyListener)) {\n    anchorEl.removeAttribute('target');\n    anchorEl.removeAttribute('onclick');\n    anchorEl.addEventListener('click', event => {\n      event.preventDefault();\n\n      const condition = !!anchorEl.href && !/^#/.test(anchorEl.getAttribute('href')) && !!/^http/.test(anchorEl.href) && !/\\.(jpe?g|png|svg|webp)$/.test(anchorEl.href) && !/remove|add|delete|(s|-|)open(s|-)|close|toggle/.test(JSON.stringify(getAttributes(anchorEl)));\n\n      if (condition) {\n        let type = 'navigate';\n        const regex = new RegExp('^" + baseUrl + "', 'g');\n        if (/^http/.test(anchorEl.href) && !regex.test(anchorEl.href)) {\n          type = 'browser';\n        }\n        window.ReactNativeWebView.postMessage(JSON.stringify({ type, payload: { url: anchorEl.href } }));\n      }\n    });\n    anchorClicks.push(anchorEl);\n  }\n}\ndocument.addEventListener('touchstart', handleTouch);\n";};

var handleScrollTop = "\n  let requestAnimationFrameId;\n\n  const handleScrollTop = () => {\n    const st = document.documentElement && document.body\n      ? document.documentElement.scrollTop || document.body.scrollTop\n      : 0;\n    if (st > 0) {\n      requestAnimationFrameId = window.requestAnimationFrame(handleScrollTop);\n      window.scrollTo(0, st - st / 5);\n    } else {\n      window.cancelAnimationFrame(requestAnimationFrameId);\n    }\n  }\n\n  window.addEventListener(\"message\", event => {\n    if (event.data) {\n      let parsedEvent = JSON.parse(event.data);\n      if (parsedEvent.type === 'scrolltop') {\n        handleScrollTop();\n      }\n    }\n  });\n";

var useWebViewLoadEnd = function (baseUrl, javascript, isScrollTop) {
  var webViewRef = useRef(null);
  var handleLoadEnd = function () {
    if (webViewRef.current) {
      webViewRef.current.injectJavaScript("\n        " + javascript + "\n        " + handleAnchorElement(baseUrl) + "\n        " + handleScrollTop + "\n      ");
    }
  };
  useEffect(function () {
    var _a;
    if (isScrollTop) {
      (_a = webViewRef.current) === null || _a === void 0 ? void 0 : _a.postMessage(JSON.stringify({ type: 'scrolltop' }));
    }
  }, [isScrollTop]);
  return {
    webViewRef: webViewRef,
    onLoadEnd: handleLoadEnd };

};

var searcher = function (str) {return function (s) {return str.search(s) !== -1;};};
var replacer = function (str) {return str.replace(/(\/|\?.*)$/g, '').replace(/^.*\//g, '');};
var checkYoutubeTypeAccount = function (url) {return searcher(url)('/user/') ? 'user' : 'chanel';};
function getUrl(url) {
  if (searcher(url)('instagram.com')) {
    return {
      ios: "instagram://user?username=" + replacer(url),
      android: "intent://instagram.com/_u/" + replacer(url) + "/#Intent;package=com.instagram.android;scheme=https;end",
      browser: url };

  } else
  if (searcher(url)('plus.google.com')) {
    return {
      ios: "gplus://plus.google.com/u/0/" + replacer(url),
      android: "gplus://plus.google.com/u/0/" + replacer(url),
      browser: url };

  } else
  if (searcher(url)(/fb:\/\/|facebook.com/g)) {
    return {
      ios: url,
      android: "intent://#Intent;package=com.facebook.katana;scheme=" + url + ";end",
      browser: !searcher(url)('fb://') ? url : "https://www.facebook.com/" + url.replace(/fb:\/\/(profile|page|group)\/(\?id=|)/g, '') };

  } else
  if (searcher(url)('twitter.com')) {
    return {
      ios: "twitter://user?screen_name=" + replacer(url),
      android: "intent://twitter.com/" + replacer(url) + "#Intent;package=com.twitter.android;scheme=https;end",
      browser: url };

  } else
  if (searcher(url)('youtube.com')) {
    return {
      ios: "vnd.youtube://www.youtube.com/" + checkYoutubeTypeAccount(url) + "/" + replacer(url),
      android: "intent://www.youtube.com/" + checkYoutubeTypeAccount(url) + "/" + replacer(url) + "#Intent;package=com.google.android.youtube;scheme=https;end",
      browser: url };

  } else
  if (searcher(url)('pinterest.com')) {
    return {
      ios: "pinterest://user/" + replacer(url),
      android: "pinterest://www.pinterest.com/" + replacer(url),
      browser: url };

  } else
  {
    return {
      ios: url,
      android: url,
      browser: url };

  }
}
function socialLinking(url) {
  return __awaiter(this, void 0, void 0, function () {
    return __generator(this, function (_a) {
      return [2 /*return*/, Linking.openURL(Platform.OS === 'ios' ? getUrl(url).ios : getUrl(url).android)["catch"](function (_) {
        openBrowserAsync(getUrl(url).browser);
      })];
    });
  });
}
function deepLinking(url) {
  return __awaiter(this, void 0, void 0, function () {
    var data, metaPropertyIos, metaPropertyAndroid, metaPropertyContent;
    return __generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          if (!searcher(url)('facebook.com')) return [3 /*break*/, 5];
          _a.label = 1;
        case 1:
          _a.trys.push([1, 3,, 4]);
          return [4 /*yield*/, axios.get(url)];
        case 2:
          data = _a.sent().data;
          metaPropertyIos = data.replace(/\/>/g, '/>\n').match(/<meta property="al:ios:url".*\/>/g)[0];
          metaPropertyAndroid = data.replace(/\/>/g, '/>\n').match(/<meta property="al:android:url".*\/>/g)[0];
          metaPropertyContent = (Platform.OS === 'ios' ? metaPropertyIos : metaPropertyAndroid).
          match(/content.*(?=")/g)[0].
          replace(/content="/g, '');
          socialLinking(metaPropertyContent);
          return [3 /*break*/, 4];
        case 3:
          _a.sent();
          socialLinking(url);
          return [3 /*break*/, 4];
        case 4:return [3 /*break*/, 6];
        case 5:
          socialLinking(url);
          _a.label = 6;
        case 6:return [2 /*return*/];}

    });
  });
}

var useWebViewMessage = function (navigation, screenName, stringParams) {
  if (stringParams === void 0) {stringParams = '';}
  var _a = useState(''),title = _a[0],setTitle = _a[1];
  var waitRef = useRef(false);
  var handleMessage = function (event) {return __awaiter(void 0, void 0, void 0, function () {
      var _a, payload, type;
      return __generator(this, function (_b) {
        _a = JSON.parse(event.nativeEvent.data), payload = _a.payload, type = _a.type;
        switch (type) {
          case 'mounted':{
              setTitle(payload.title);
              break;
            }
          case 'browser':{
              deepLinking(payload.url);
              break;
            }
          case 'navigate':{
              if (!waitRef.current) {
                waitRef.current = true;
                navigation.push(screenName, {
                  url: "" + payload.url + (!!stringParams ? payload.url.includes('?') ? '&' : '?' : '') + stringParams });

                setTimeout(function () {
                  waitRef.current = false;
                }, 1000);
              }
              break;
            }
}

        return [2 /*return*/];
      });
    });};
  return {
    title: title,
    onMessage: handleMessage };

};

var isIphoneX = function () {
  var _a = Dimensions.get('window'),width = _a.width,height = _a.height;
  return Platform.OS === 'ios' && !Platform.isPad && !Platform.isTVOS && (height === 812 || width === 812 || height === 896 || width === 896);
};
var bottomBarHeight = isIphoneX() ? 23 : 0;

// export const darks = {
//   dark1: '#F1F1F1',
//   dark2: '#E0E0E0',
//   dark3: '#8F8F8F',
//   dark4: '#595959',
//   gray1: '#51514E',
//   gray2: '#44443F',
//   gray3: '#3C3C3A',
//   gray4: '#343434',
//   gray5: '#343433',
//   light: '#313131',
// };

var webviewCss = "\n  * {\n    user-select: none !important;\n  }\n  body {\n    padding-bottom: " + bottomBarHeight + "px;\n  }\n  html {\n    margin-top: 0 !important;\n  }\n  .wil-header-wrap, [class^=\"wil-footer\"] {\n    display: none;\n  }\n  #shopify-section-header,\n  #shopify-section-footer,\n  #header,\n  #footer {\n    display: none !important;\n  }\n";

var MagicContentComponent = function (_a) {
  var navigation = _a.navigation,baseUrl = _a.baseUrl,url = _a.url,screenName = _a.screenName,_b = _a.stringParams,stringParams = _b === void 0 ? '' : _b,_c = _a.css,css = _c === void 0 ? '' : _c,_d = _a.javascript,javascript = _d === void 0 ? '' : _d,_e = _a.renderLoading,renderLoading = _e === void 0 ? function () {return /*#__PURE__*/React.createElement(ActivityIndicator, { size: "small" });} : _e,_f = _a.showsVerticalScrollIndicator,showsVerticalScrollIndicator = _f === void 0 ? false : _f,_g = _a.showsHorizontalScrollIndicator,showsHorizontalScrollIndicator = _g === void 0 ? false : _g,_h = _a.pullToRefreshEnabled,pullToRefreshEnabled = _h === void 0 ? true : _h,_j = _a.allowsFullscreenVideo,allowsFullscreenVideo = _j === void 0 ? true : _j,_k = _a.sharedCookiesEnabled,sharedCookiesEnabled = _k === void 0 ? true : _k,_l = _a.isScrollTop,isScrollTop = _l === void 0 ? false : _l,rest = __rest(_a, ["navigation", "baseUrl", "url", "screenName", "stringParams", "css", "javascript", "renderLoading", "showsVerticalScrollIndicator", "showsHorizontalScrollIndicator", "pullToRefreshEnabled", "allowsFullscreenVideo", "sharedCookiesEnabled", "isScrollTop"]);
  var _m = useState(false),loading = _m[0],setLoading = _m[1];
  var onMessage = useWebViewMessage(navigation, screenName, stringParams).onMessage;
  var _o = useWebViewLoadEnd(baseUrl, javascript, isScrollTop),onLoadEnd = _o.onLoadEnd,webViewRef = _o.webViewRef;
  var _css = webviewCss + css;
  var handleLoadStart = function () {
    setLoading(true);
  };
  var handleLoadEnd = function (event) {
    var _a;
    setLoading(false);
    onLoadEnd();
    (_a = rest.onLoadEnd) === null || _a === void 0 ? void 0 : _a.call(rest, event);
  };
  var handleMessage = function (event) {
    var _a;
    onMessage(event);
    (_a = rest.onMessage) === null || _a === void 0 ? void 0 : _a.call(rest, event);
  };
  return /*#__PURE__*/React.createElement(View, { flex: true, tachyons: "relative" },
  loading && /*#__PURE__*/React.createElement(View, { style: { backgroundColor: 'rgba(255,255,255,0.4)' }, tachyons: ['absolute', 'absoluteFill', 'z5', 'flex', 'justifyCenter', 'itemsCenter'] }, renderLoading()), /*#__PURE__*/
  React.createElement(NavigationSuspense, null, /*#__PURE__*/
  React.createElement(WebView, __assign({}, rest, { decelerationRate: "normal", ref: webViewRef, source: { uri: url }, style: {
      width: '100%',
      height: '100%' },
    originWhitelist: ['*'], showsVerticalScrollIndicator: showsVerticalScrollIndicator, showsHorizontalScrollIndicator: showsHorizontalScrollIndicator, javaScriptEnabled: true, pullToRefreshEnabled: pullToRefreshEnabled, allowsFullscreenVideo: allowsFullscreenVideo, sharedCookiesEnabled: sharedCookiesEnabled, injectedJavaScript: "\n            document.head.insertAdjacentHTML('beforeend', '<style>" + _css.replace(/\n/g, '').replace(/\s+/g, ' ') + "</style>');\n            window.ReactNativeWebView.postMessage(JSON.stringify({ type: 'mounted', payload: { title: document.title } }));\n          ", onLoadStart: handleLoadStart, onLoadEnd: handleLoadEnd, onMessage: handleMessage }))));
};
var MagicContent = withNavigation(MagicContentComponent);

var CartTotal = function (_a) {
  var cartUri = _a.cartUri,onFetchCartTotalEnd = _a.onFetchCartTotalEnd;
  var webViewRef = useRef(null);
  var _b = useState(false),reload = _b[0],setReload = _b[1];
  var intervalRef = useRef(0);
  var cartTotal = 0;
  useEffect(function () {
    intervalRef.current = setInterval(function () {
      setReload(function (reload) {return !reload;});
    }, 5000);
    return function () {
      clearInterval(intervalRef.current);
    };
  }, [reload]);
  if (reload) {
    return null;
  }
  return /*#__PURE__*/React.createElement(View, { style: { position: 'absolute' } }, /*#__PURE__*/
  React.createElement(WebView, { source: { uri: cartUri }, ref: webViewRef, style: { width: 0, height: 0 }, onLoadEnd: function () {
      var _a;
      (_a = webViewRef.current) === null || _a === void 0 ? void 0 : _a.injectJavaScript("\n            const qtyEls = document.querySelectorAll('form[method=\"post\"] input[type=\"number\"]');\n            if (qtyEls.length > 0) {\n              qtyEls.forEach(qtyEl => {\n                window.ReactNativeWebView.postMessage(JSON.stringify({ qtyValue: qtyEl.value || 0 }));\n              })\n            } else {\n              window.ReactNativeWebView.postMessage(JSON.stringify({ qtyValue: 0 }));\n            }\n          ");
    }, onMessage: function (event) {
      var qtyValue = JSON.parse(event.nativeEvent.data).qtyValue;
      if (qtyValue === 0) {
        cartTotal = 0;
      } else
      {
        cartTotal += Number(qtyValue);
      }
      onFetchCartTotalEnd(cartTotal);
    } }));
};
var ECartTotal = /*#__PURE__*/memo(CartTotal);

var VALUE = 100;
function getInputRange(outputRange) {
  return outputRange.map(function (_item, index) {
    return VALUE * index / (outputRange.length - 1);
  });
}
function useFlyAnimation(_a) {
  var coordinates = _a.coordinates,duration = _a.duration;
  var anim = useRef(new Animated.Value(0));
  var outputRangeX = Array.from(new Set(coordinates.map(function (item) {return item.x;})));
  var outputRangeY = Array.from(new Set(coordinates.map(function (item) {return item.y;})));
  var onFly = useCallback(function () {
    Animated.timing(anim.current, {
      toValue: VALUE,
      duration: duration,
      useNativeDriver: true }).
    start();
  }, [duration]);
  var inputRangeX = getInputRange(outputRangeX);
  var inputRangeY = getInputRange(outputRangeY);
  var animatedX = anim.current.interpolate({
    inputRange: inputRangeX,
    outputRange: outputRangeX,
    extrapolate: 'clamp' });

  var animatedY = anim.current.interpolate({
    inputRange: inputRangeY,
    outputRange: outputRangeY,
    extrapolate: 'clamp' });

  return { animatedX: animatedX, animatedY: animatedY, onFly: onFly };
}

function useAnimated(_a) {
  var toValue = _a.toValue,_b = _a.duration,duration = _b === void 0 ? 200 : _b,_c = _a.useNativeDriver,useNativeDriver = _c === void 0 ? true : _c;
  var animation = useRef(new Animated.Value(0));
  var onAnimated = useCallback(function () {
    Animated.timing(animation.current, {
      toValue: toValue,
      duration: duration,
      useNativeDriver: useNativeDriver }).
    start();
  }, [duration, toValue, useNativeDriver]);
  return [animation.current, onAnimated];
}

function getError(field, value) {
  var error = '';
  if (typeof value !== 'string' && !Array.isArray(value)) {
    return '';
  }
  if (field.required) {
    error = value ? '' : 'required';
  }
  if (field.minLength) {
    error = field.required && !value || value.length >= field.minLength ? error : 'minLength';
  }
  if (field.maxLength) {
    error = value.length <= field.maxLength ? error : 'maxLength';
  }
  if (field.pattern) {
    error =
    field.required && !value || value.length < (field.minLength || 0) || value.length > (field.maxLength || Infinity) || field.pattern.test(value) ?
    error :
    'pattern';
  }
  return error;
}
function useForm() {
  var _a = useState({}),result = _a[0],setResult = _a[1];
  var _b = useState({}),errors = _b[0],setErrors = _b[1];
  var resultRef = useRef({});
  var errorsRef = useRef({});
  var getNewResult = function (result) {return Object.values(result).length ? result : resultRef.current;};
  var getNewErrors = function (errors) {return Object.values(errors).length ? errors : errorsRef.current;};
  var onChange = function (field) {
    var _a, _b;
    resultRef.current = __assign(__assign({}, resultRef.current), (_a = {}, _a[field.name] = field.defaultValue, _a));
    errorsRef.current = __assign(__assign({}, errorsRef.current), (_b = {}, _b[field.name] = getError(field, field.defaultValue), _b));
    return function (value) {
      setResult(function (result) {
        var _a;
        return __assign(__assign(__assign({}, resultRef.current), result), (_a = {}, _a[field.name] = value, _a));
      });
      setErrors(function (errors) {
        var _a;
        return __assign(__assign(__assign({}, errorsRef.current), errors), (_a = {}, _a[field.name] = getError(field, value), _a));
      });
    };
  };
  var onSubmit = function (handleSubmit) {return function (event) {
      event.preventDefault && event.preventDefault();
      setResult(getNewResult);
      setErrors(getNewErrors);
      handleSubmit({
        result: getNewResult(result),
        errors: getNewErrors(errors),
        isError: Object.values(getNewErrors(errors)).reduce(function (str, item) {return str + item;}, '') !== '' });

    };};
  return {
    onChange: onChange,
    onSubmit: onSubmit,
    errors: errors };

}

function configureLayoutAnimation(event) {
  return LayoutAnimation.configureNext({
    duration: isIOS ? event.duration : 300,
    create: {
      type: isIOS ? LayoutAnimation.Types.keyboard : LayoutAnimation.Types.easeInEaseOut,
      property: LayoutAnimation.Properties.opacity },

    update: {
      type: isIOS ? LayoutAnimation.Types.keyboard : LayoutAnimation.Types.easeInEaseOut } });


}
function useKeyboardHeightAnimation() {
  var _a = useState(0),keyboardHeight = _a[0],setKeyBoardHeight = _a[1];
  var switchEventNameShow = isIOS ? 'keyboardWillShow' : 'keyboardDidShow';
  var switchEventNameHide = isIOS ? 'keyboardWillHide' : 'keyboardDidHide';
  var handleKeyboardShow = Keyboard.addListener(switchEventNameShow, function (event) {
    var keyboardHeight = isIOS ? event.endCoordinates.height : event.endCoordinates.height - 100;
    configureLayoutAnimation(event);
    setKeyBoardHeight(keyboardHeight);
  });
  var handleKeyboardHide = Keyboard.addListener(switchEventNameHide, function (event) {
    configureLayoutAnimation(event);
    setKeyBoardHeight(0);
  });
  useUnmount(function () {
    handleKeyboardShow.remove();
    handleKeyboardHide.remove();
  });
  return keyboardHeight;
}

export { Button, Card, Container, Divider, ECartTotal, FlatList, FormItem, HeaderBase, Icons, Image, Input, InputBase, MagicContent, Modal, ModalBase, NavigationSuspense, OfflineNotice, Popover, SCREEN_HEIGHT, SCREEN_WIDTH, SelectList, Text, TextDivider, TextInputMentions, ThemeConsumer, ThemeProvider, Toast, View, isIpad, isSmallDevice, tachyonsStyles as tachyons, tailwindStyles as tailwind, useAnimated, useFlyAnimation, useForm, useKeyboardHeightAnimation, useMeasure, useMount, useSelectList, useSwitchAnimation, useTheme, useToggle, useUnmount, withTachyons, withTailwind, withTextStyles, withTheme, withViewStyles };
